<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>A tutorial to support multivariate Bayesian analyses in nursing research - 4&nbsp; mvBIMBAM</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Troubleshooting.html" rel="next">
<link href="./bnlearn.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./mvbimbam.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">mvBIMBAM</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">A tutorial to support multivariate Bayesian analyses in nursing research</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prep.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bnlearn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">bnlearn</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mvbimbam.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">mvBIMBAM</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Troubleshooting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Troubleshooting mvBIMBAM installation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./InstallationHack.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">mvBIMBAM Installation Hack</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview"><span class="header-section-number">4.1</span> Overview</a>
  <ul class="collapse">
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements"><span class="header-section-number">4.1.1</span> Acknowledgements</a></li>
  <li><a href="#mvbimbam" id="toc-mvbimbam" class="nav-link" data-scroll-target="#mvbimbam"><span class="header-section-number">4.1.2</span> mvBIMBAM</a></li>
  </ul></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries"><span class="header-section-number">4.2</span> Load libraries</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="header-section-number">4.3</span> Data</a></li>
  <li><a href="#mvbimbam-analyses" id="toc-mvbimbam-analyses" class="nav-link" data-scroll-target="#mvbimbam-analyses"><span class="header-section-number">4.4</span> mvBIMBAM analyses</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"><span class="header-section-number">4.5</span> Results</a>
  <ul class="collapse">
  <li><a href="#bayes-factors-bf" id="toc-bayes-factors-bf" class="nav-link" data-scroll-target="#bayes-factors-bf"><span class="header-section-number">4.5.1</span> Bayes factors (BF)</a></li>
  <li><a href="#bayesian-posterior-probabilities-of-association" id="toc-bayesian-posterior-probabilities-of-association" class="nav-link" data-scroll-target="#bayesian-posterior-probabilities-of-association"><span class="header-section-number">4.5.2</span> Bayesian posterior probabilities of association</a></li>
  </ul></li>
  <li><a href="#save-data-and-results" id="toc-save-data-and-results" class="nav-link" data-scroll-target="#save-data-and-results"><span class="header-section-number">4.6</span> Save data and results</a>
  <ul class="collapse">
  <li><a href="#mvbimbam-bfs" id="toc-mvbimbam-bfs" class="nav-link" data-scroll-target="#mvbimbam-bfs"><span class="header-section-number">4.6.1</span> mvBIMBAM BFs</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion"><span class="header-section-number">4.7</span> Discussion</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">4.8</span> Conclusion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">mvBIMBAM</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="overview"><span class="header-section-number">4.1</span> Overview</h2>
<p>Note this code is dependent upon code from the <a href="./prep.html">data preparation</a> section which creates the mvBIMBAM input files detailed below.</p>
<section id="acknowledgements" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="acknowledgements"><span class="header-section-number">4.1.1</span> Acknowledgements</h3>
<p>Please note that this code was adapted from previous work led by Jerry Zhang, Lacey Heinsberg, and Daniel Weeks:</p>
<p>Zhang JZ, Heinsberg LW, Krishnan M, Hawley NL, Major TJ, Carlson JC, Harré Hindmarsh J, Watson H, Qasim M, Stamp LK, Dalbeth N, Murphy R, Sun G, Cheng H, Naseri T, Reupena MS, Kershaw EE, Deka R, McGarvey ST, Minster RL, Merriman TR, Weeks DE. Multivariate analysis of a missense variant in CREBRF reveals associations with measures of adiposity in people of Polynesian ancestries. Genet Epidemiol. 2023 Feb;47(1):105-118. doi: 10.1002/gepi.22508. Epub 2022 Nov 9. PMID: 36352773; PMCID: PMC9892232.</p>
<p><a href="https://github.com/lwheinsberg/mvCREBRF">GitHub Repository</a></p>
</section>
<section id="mvbimbam" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2" class="anchored" data-anchor-id="mvbimbam"><span class="header-section-number">4.1.2</span> mvBIMBAM</h3>
<p>mvBIMBAM is a command line program for multivariate genetic association analysis of multiple related phenotypes. In the mvBIMBAM framework, a global null model representing no association between phenotypes and genotypes is compared with an exhaustive combination of alternative models in which all different combinations of phenotype-genotype associations are considered (see the <code>mph 2</code> analysis parameter details below). For the alternative models, the mvBIMBAM methodology splits phenotype-genotype associations into all possible partitions of U, D, and I, each representing ‘unassociated’, ‘directly’, and ‘indirectly’ associated. Details about the analyses and interpretation can be found below.</p>
<p>NOTE: The software executable is called ‘bimbam’, but we refer to the program as ‘mvBIMBAM’ throughout.</p>
</section>
</section>
<section id="load-libraries" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="load-libraries"><span class="header-section-number">4.2</span> Load libraries</h2>
<p>Load the libraries needed to run the code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>select <span class="ot">=</span> dplyr<span class="sc">::</span>select</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The 'preprocessCore' is a Bioconductor package</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(preprocessCore)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pander)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="data"><span class="header-section-number">4.3</span> Data</h2>
<p>Most of our data (created in the <a href="./prep.html">data preparation</a> section) are directly analyzed by the mvBIMBAM command line program (see below) and not formerly loaded into R. Here though, we will read in our stored lists of phenotypes/variants of interest. See the <a href="./intro.html">introduction section</a> for information about phenotypes and SNPs of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"./data/TraitsGenes.RData"</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>) <span class="do">###CUSTOMIZE** (if input file name was changed in 00_ workflow)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading objects:
  traits
  genes
  trait_mapping
  custom_labels
  custom_labels2
  df_vertex_table</code></pre>
</div>
</div>
<p>Could also hard code these variables as shown below.</p>
<pre><code>traits &lt;- c("EMO_tscore", "bdito", "FAT_tscore", "paohcif", "EPSscore", "pain") ###CUSTOMIZE**
trait_mapping &lt;- c("Anxiety", "Depression", "Fatigue", "Cognitive function", "Sleepiness", "Pain") ###CUSTOMIZE**
custom_labels &lt;- setNames(trait_mapping, traits)
custom_labels2 &lt;- c(custom_labels, setNames(genes, genes))
genes &lt;- c("rs4880", "rs5746136", "rs1041740", "rs10432782", "rs4135225", "rs7522705") ###CUSTOMIZE**</code></pre>
</section>
<section id="mvbimbam-analyses" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="mvbimbam-analyses"><span class="header-section-number">4.4</span> mvBIMBAM analyses</h2>
<p>We will now apply mvBIMBAM to the synthetic data set we are working with. Rather than navigating the analyses in the terminal, we can make our code reproducible in R and call mvBIMBAM through the R using <code>call system()</code>. More information about the program can be found under the original <a href="https://github.com/heejungshim/mvBIMBAM/blob/master/doc/mvBIMBAM_Manual.pdf">mvBIMBAM documentation</a> on GitHub. In brief, the command we need contains the following parameters:</p>
<p>-g: location of the genotype data (for our example the location is: ./inputs/geno_bimbam.txt)<br>
-p: location of the phenotype data (for our example the location is: ./inputs/pheno_bimbam.txt)<br>
-o: prefix for output files (for our example we will use: bimbam_out)<br>
-f: number of phenotypes to analyze (length(traits), for our example this number is 6)<br>
-A: indicate multiple values for sigma (see mvBIMBAM manual for more details; for our example we will use a range of values from 0.05 to 0.4)<br>
-mph 1 vs.&nbsp;mph 2: analysis settings (see mvBIMBMAM manual for more details and our note below; for this example we will run mph 2)<br>
-r: random seed for reproducibility of results**</p>
<p>**Random seed: The mvBIMBAM article states: “We maximized the likelihood using an EM algorithm; four independent runs of the algorithm from different starting points produced essentially identical results.” This appears to be implemented in a random way, and so depends on the random number seed. For the purposes of this tutorial, we set a seed number so that the results would be identical each time. In practice, this step is not necessary, so the “-r 5254” section can be deleted from the code below.</p>
<p>A note about <code>mph 1</code> vs.&nbsp;<code>mph 2</code>: In brief, <code>-mph 1</code> performs a simple multivariate test of the null hypothesis vs a general multivariate alternative. Because it does not consider all partitions of the alternative hypothesis, it is quite fast, can handle a moderately large number of phenotypes, and be run on a whole-genome scale. However, it was originally intended to be used only as needed in an initial filtering step before running the more computationally-intensive <code>-mph 2</code> analyses. Because of the way <code>-mph 1</code> functions, it does not allow you to ask interesting questions such as which phenotypes are associated with each SNP. To do this more detailed analysis use the <code>-mph 2</code> option. Note that if you have only a handful of phenotypes, you may be able to run the <code>-mph 2</code> analysis genome wide on all variants and skip <code>-mph 1</code> all together.</p>
<p>A quick warning from the creators of mvBIMBAM: the prior used in mvBIMBAM is based on the idea that if a variant is associated with one phenotype, then it is likely associated with multiple phenotypes. That is, it is relatively permissive of associations with multiple phenotypes, and does not attempt to be “skeptical” of additional associations. In this sense it is more suited to hypothesis generation than of hypothesis testing. The prior can be changed to be more skeptical, but this adjustment has not been explored in detail as it is beyond the scope of this tutorial.</p>
<p>We now use the ‘system’ command to run the mvBIMBAM program on our data set, which outputs the following messages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call system() to run mvBIMBAM by specifying settings for analysis using </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the following settings:  </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>call <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"bimbam -g ./inputs/geno_bimbam.txt -p ./inputs/pheno_bimbam.txt -o bimbam_out -mph 2 -f "</span>, <span class="fu">length</span>(traits), <span class="st">" -A 0.05 -A 0.1 -A 0.2 -A 0.4 -r 5254"</span>) </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="do">###CUSTOMIZE** (call object, do so if input file names (i.e., geno_bimbam.txt or pheno_bimbam.txt) were changed in 00_ workflow)</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="do">###CUSTOMIZE** (optional, can change output file name from bimbam_out to a name of your choice)</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(call)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "bimbam -g ./inputs/geno_bimbam.txt -p ./inputs/pheno_bimbam.txt -o bimbam_out -mph 2 -f 6 -A 0.05 -A 0.1 -A 0.2 -A 0.4 -r 5254"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Please note that this can take a few minutes</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(call, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "-bimbam: file 0 has 763 individual and 6 snps"   
[2] "-bimbam: read file 0 again "                     
[3] "-bimbam: number of phenotypes = 6"               
[4] "total = 729"                                     
[5] "output/bimbam_out.mph.txt has been created."     
[6] "output/bimbam_out.mph.prob.txt has been created."
[7] "output/bimbam_out.mph.BFs.txt has been created." 
attr(,"status")
[1] 1</code></pre>
</div>
</div>
<pre><code>bimbam -g ./inputs/geno_bimbam.txt -p ./inputs/pheno_bimbam.txt 
-o bimbam_out -mph 2 -f 6 -A 0.05 -A 0.1 -A 0.2 -A 0.4 -r 5254</code></pre>
<p>Once we run the chunk above, we can then find our output at ./output/bimbam_out.mph.BFs.txt.</p>
</section>
<section id="results" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="results"><span class="header-section-number">4.5</span> Results</h2>
<section id="bayes-factors-bf" class="level3" data-number="4.5.1">
<h3 data-number="4.5.1" class="anchored" data-anchor-id="bayes-factors-bf"><span class="header-section-number">4.5.1</span> Bayes factors (BF)</h3>
<p>The evidence against the null hypothesis is the sum of Bayes factors (BF) (log10 scale) of all partitions weighted by a diffuse prior. Let’s read in our BF factor results file to take a look.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">pillar.sigfig =</span> <span class="dv">10</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"./output/bimbam_out.mph.BFs.txt"</span>, <span class="at">col_names =</span> <span class="cn">FALSE</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="do">###CUSTOMIZE** (if output file names are changed in the call object above)</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> m1 <span class="sc">%&gt;%</span> <span class="fu">select_if</span>(<span class="sc">~</span><span class="fu">any</span>(<span class="sc">!</span><span class="fu">is.na</span>(.)))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(m1) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"SNP"</span>, <span class="st">"BF"</span>, <span class="st">"BF_All_Partition"</span>, trait_mapping)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(m1)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> m1 <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>BF_All_Partition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View results</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pander</span>(m1, <span class="at">digits =</span> <span class="dv">4</span>, <span class="at">caption =</span> <span class="st">"Bayes Factors"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Bayes Factors (continued below)</caption>
<colgroup>
<col style="width: 17%">
<col style="width: 10%">
<col style="width: 13%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 28%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">SNP</th>
<th style="text-align: center;">BF</th>
<th style="text-align: center;">Anxiety</th>
<th style="text-align: center;">Depression</th>
<th style="text-align: center;">Fatigue</th>
<th style="text-align: center;">Cognitive.function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">rs4880</td>
<td style="text-align: center;">11.64</td>
<td style="text-align: center;">-0.2122</td>
<td style="text-align: center;">0.9347</td>
<td style="text-align: center;">-0.405</td>
<td style="text-align: center;">2.041</td>
</tr>
<tr class="even">
<td style="text-align: center;">rs5746136</td>
<td style="text-align: center;">8.751</td>
<td style="text-align: center;">0.1084</td>
<td style="text-align: center;">-0.3045</td>
<td style="text-align: center;">0.6435</td>
<td style="text-align: center;">0.2681</td>
</tr>
<tr class="odd">
<td style="text-align: center;">rs1041740</td>
<td style="text-align: center;">13.69</td>
<td style="text-align: center;">12.52</td>
<td style="text-align: center;">0.6702</td>
<td style="text-align: center;">4.629</td>
<td style="text-align: center;">6.156</td>
</tr>
<tr class="even">
<td style="text-align: center;">rs10432782</td>
<td style="text-align: center;">4.878</td>
<td style="text-align: center;">1.619</td>
<td style="text-align: center;">1.85</td>
<td style="text-align: center;">-0.3186</td>
<td style="text-align: center;">0.05509</td>
</tr>
<tr class="odd">
<td style="text-align: center;">rs4135225</td>
<td style="text-align: center;">32.48</td>
<td style="text-align: center;">0.6325</td>
<td style="text-align: center;">1.603</td>
<td style="text-align: center;">25.11</td>
<td style="text-align: center;">5.586</td>
</tr>
<tr class="even">
<td style="text-align: center;">rs7522705</td>
<td style="text-align: center;">1.672</td>
<td style="text-align: center;">-0.2782</td>
<td style="text-align: center;">-0.2978</td>
<td style="text-align: center;">-0.2944</td>
<td style="text-align: center;">0.5971</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Sleepiness</th>
<th style="text-align: center;">Pain</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0.02955</td>
<td style="text-align: center;">3.437</td>
</tr>
<tr class="even">
<td style="text-align: center;">6.421</td>
<td style="text-align: center;">0.6853</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.00889</td>
<td style="text-align: center;">-0.2024</td>
</tr>
<tr class="even">
<td style="text-align: center;">-0.05979</td>
<td style="text-align: center;">-0.1757</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4.684</td>
<td style="text-align: center;">11.63</td>
</tr>
<tr class="even">
<td style="text-align: center;">-0.3565</td>
<td style="text-align: center;">0.2123</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Interpretation: The above table presents the overall log10 BF for the network, and individual log10 BFs for each trait. Strong evidence of association is defined as log10 BF &gt; 5; suggestive evidence is defined as 1.5 &lt; log10 BF &lt; 5; and negligible evidence is defined as log10 BF &lt; 1.5.</p>
<p>In this synthetic data set, we see that all SNPs (rs7522705 less so and rs4135225 more so) appear to play a role in this network. rs4880 has suggestive associations with cognitive function and pain; rs5746136 has significant associations with sleepiness; rs1041740 has significant associations with anxiety and cognitive function and a suggestive association with fatigue, etc.</p>
</section>
<section id="bayesian-posterior-probabilities-of-association" class="level3" data-number="4.5.2">
<h3 data-number="4.5.2" class="anchored" data-anchor-id="bayesian-posterior-probabilities-of-association"><span class="header-section-number">4.5.2</span> Bayesian posterior probabilities of association</h3>
<p>In addition to log10 BFs, probabilities for no association, direct association, and indirect association are output while marginal posterior probabilities of association (MPPA) are calculated by summing the marginal posterior probabilities of direct and indirect association.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in second result file</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">"./output/bimbam_out.mph.prob.txt"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize an empty list to store the matrix for each rs#</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>matrix_list <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">"list"</span>, <span class="fu">length</span>(s))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Process each line</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(s)) {</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Split the line by spaces</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  split_values <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(s[i], <span class="st">"</span><span class="sc">\\</span><span class="st">s+"</span>)[[<span class="dv">1</span>]]</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Store the rs# value</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  rs <span class="ot">&lt;-</span> split_values[<span class="dv">1</span>]</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  m2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">na.omit</span>(<span class="fu">as.numeric</span>(<span class="fu">str_split</span>(s, <span class="st">" "</span>)[[i]])), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  m2 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(m2, <span class="dv">1</span> <span class="sc">-</span> (m2[<span class="dv">1</span>,] <span class="sc">+</span> m2[<span class="dv">2</span>,]))</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set the column and row names</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(m2) <span class="ot">&lt;-</span> trait_mapping</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(m2) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(rs, <span class="st">"_Unassociated"</span>), <span class="fu">paste0</span>(rs, <span class="st">"_Directly"</span>), <span class="fu">paste0</span>(rs, <span class="st">"_Indirectly"</span>))</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Store the matrix in the list, using rs# as the name</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  matrix_list[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">rs =</span> rs, <span class="at">m2 =</span> m2)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all matrices into a single matrix</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>m2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(matrix_list, <span class="cf">function</span>(x) x<span class="sc">$</span>m))</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(m2)) {</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  m2[, i] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\"</span><span class="st">"</span>, <span class="st">""</span>, m2[, i]))</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pander</span>(m2, <span class="at">digits =</span> <span class="dv">4</span>, <span class="at">caption =</span> <span class="st">"Bayesian Probabilities"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Bayesian Probabilities (continued below)</caption>
<colgroup>
<col style="width: 41%">
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">Anxiety</th>
<th style="text-align: center;">Depression</th>
<th style="text-align: center;">Fatigue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>rs4880_Unassociated</strong></td>
<td style="text-align: center;">0.3751</td>
<td style="text-align: center;">0.0881</td>
<td style="text-align: center;">0.3875</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs4880_Directly</strong></td>
<td style="text-align: center;">0.3415</td>
<td style="text-align: center;">0.9115</td>
<td style="text-align: center;">0.2468</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4880_Indirectly</strong></td>
<td style="text-align: center;">0.2834</td>
<td style="text-align: center;">0.00035</td>
<td style="text-align: center;">0.3657</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs5746136_Unassociated</strong></td>
<td style="text-align: center;">0.1947</td>
<td style="text-align: center;">0.3647</td>
<td style="text-align: center;">0.07658</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs5746136_Directly</strong></td>
<td style="text-align: center;">0.8037</td>
<td style="text-align: center;">0.5952</td>
<td style="text-align: center;">0.9207</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs5746136_Indirectly</strong></td>
<td style="text-align: center;">0.00154</td>
<td style="text-align: center;">0.04015</td>
<td style="text-align: center;">0.00272</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs1041740_Unassociated</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.07304</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs1041740_Directly</strong></td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.9007</td>
<td style="text-align: center;">0.4676</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs1041740_Indirectly</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.02628</td>
<td style="text-align: center;">0.5324</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs10432782_Unassociated</strong></td>
<td style="text-align: center;">0.00504</td>
<td style="text-align: center;">0.00493</td>
<td style="text-align: center;">0.3862</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs10432782_Directly</strong></td>
<td style="text-align: center;">0.9831</td>
<td style="text-align: center;">0.9789</td>
<td style="text-align: center;">0.3531</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs10432782_Indirectly</strong></td>
<td style="text-align: center;">0.01188</td>
<td style="text-align: center;">0.01618</td>
<td style="text-align: center;">0.2607</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4135225_Unassociated</strong></td>
<td style="text-align: center;">0.08721</td>
<td style="text-align: center;">0.01237</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs4135225_Directly</strong></td>
<td style="text-align: center;">0.9128</td>
<td style="text-align: center;">0.532</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4135225_Indirectly</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.4556</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs7522705_Unassociated</strong></td>
<td style="text-align: center;">0.4168</td>
<td style="text-align: center;">0.54</td>
<td style="text-align: center;">0.4684</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs7522705_Directly</strong></td>
<td style="text-align: center;">0.458</td>
<td style="text-align: center;">0.4192</td>
<td style="text-align: center;">0.5191</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs7522705_Indirectly</strong></td>
<td style="text-align: center;">0.1253</td>
<td style="text-align: center;">0.04077</td>
<td style="text-align: center;">0.01252</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 40%">
<col style="width: 28%">
<col style="width: 17%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">Cognitive function</th>
<th style="text-align: center;">Sleepiness</th>
<th style="text-align: center;">Pain</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>rs4880_Unassociated</strong></td>
<td style="text-align: center;">0.00077</td>
<td style="text-align: center;">0.2343</td>
<td style="text-align: center;">0.00035</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs4880_Directly</strong></td>
<td style="text-align: center;">0.9992</td>
<td style="text-align: center;">0.358</td>
<td style="text-align: center;">0.9997</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4880_Indirectly</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.4077</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs5746136_Unassociated</strong></td>
<td style="text-align: center;">0.1642</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.05097</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs5746136_Directly</strong></td>
<td style="text-align: center;">0.8279</td>
<td style="text-align: center;">0.9999</td>
<td style="text-align: center;">0.5256</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs5746136_Indirectly</strong></td>
<td style="text-align: center;">0.00791</td>
<td style="text-align: center;">9e-05</td>
<td style="text-align: center;">0.4234</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs1041740_Unassociated</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0.2551</td>
<td style="text-align: center;">0.2203</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs1041740_Directly</strong></td>
<td style="text-align: center;">0.9235</td>
<td style="text-align: center;">0.7228</td>
<td style="text-align: center;">0.3356</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs1041740_Indirectly</strong></td>
<td style="text-align: center;">0.0765</td>
<td style="text-align: center;">0.02212</td>
<td style="text-align: center;">0.4441</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs10432782_Unassociated</strong></td>
<td style="text-align: center;">0.3096</td>
<td style="text-align: center;">0.2886</td>
<td style="text-align: center;">0.3648</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs10432782_Directly</strong></td>
<td style="text-align: center;">0.6901</td>
<td style="text-align: center;">0.4881</td>
<td style="text-align: center;">0.4547</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs10432782_Indirectly</strong></td>
<td style="text-align: center;">0.00032</td>
<td style="text-align: center;">0.2233</td>
<td style="text-align: center;">0.1805</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4135225_Unassociated</strong></td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs4135225_Directly</strong></td>
<td style="text-align: center;">0.361</td>
<td style="text-align: center;">0.3242</td>
<td style="text-align: center;">0.9953</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs4135225_Indirectly</strong></td>
<td style="text-align: center;">0.639</td>
<td style="text-align: center;">0.6758</td>
<td style="text-align: center;">0.00466</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs7522705_Unassociated</strong></td>
<td style="text-align: center;">0.1034</td>
<td style="text-align: center;">0.5502</td>
<td style="text-align: center;">0.3012</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>rs7522705_Directly</strong></td>
<td style="text-align: center;">0.8893</td>
<td style="text-align: center;">0.272</td>
<td style="text-align: center;">0.6018</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>rs7522705_Indirectly</strong></td>
<td style="text-align: center;">0.00732</td>
<td style="text-align: center;">0.1778</td>
<td style="text-align: center;">0.09697</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Interpretation: The numbers above can be interpreted as the probability of no association, direct association, or indirect association, which together sum to 1 (i.e., 100%). Both directly and indirectly associated phenotypes are associated with a given genotype, but indirectly associated phenotypes are conditionally independent of the genotype given the presence of a directly associated phenotype in the model.</p>
<p>For example, in this synthetic data set, there is a suggested 100% probability that there is a direct association between rs1041740 and anxiety. In addition, while there is a 0% probability of no association between rs4135225 and sleepiness, the likelihood of direct vs.&nbsp;indirect effects is more split at 32.417% and 67.583%, respectively. While mvBIMBAM doesn’t inform of us what variable indirect associations are conditional on, we can look back at our bnlearn network for more information.</p>
</section>
</section>
<section id="save-data-and-results" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="save-data-and-results"><span class="header-section-number">4.6</span> Save data and results</h2>
<p>To conclude, save the Bayes factors from the mvBIMBAM results.</p>
<section id="mvbimbam-bfs" class="level3" data-number="4.6.1">
<h3 data-number="4.6.1" class="anchored" data-anchor-id="mvbimbam-bfs"><span class="header-section-number">4.6.1</span> mvBIMBAM BFs</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(m1, m2, <span class="at">file =</span> <span class="st">"data/mvBimBam_Results.RDdata"</span>) <span class="do">###CUSTOMIZE** (optional, can change file name of results)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="discussion" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="discussion"><span class="header-section-number">4.7</span> Discussion</h2>
<p>Note that mvBIMBAM’s default priors are “loose” (the phrase used in the software documentation), meaning associations are detected more liberally. This aligns with our results, where mvBIMBAM identified rs7522705 as suggestively important to the network (1.5 &lt; BF &lt; 5), but none of the individual phenotypes themselves were even suggestive (e.g., all BFs &lt; 1.5), while bnlearn omitted it completely. It’s worth mentioning that you can modify the default priors to be more conservative (see software documentation), but this is beyond the scope of this guide.</p>
</section>
<section id="conclusion" class="level2" data-number="4.8">
<h2 data-number="4.8" class="anchored" data-anchor-id="conclusion"><span class="header-section-number">4.8</span> Conclusion</h2>
<p>And with that, we conclude our <code>mvBIMBAM</code> tutorial! We hope that this code is documented in enough detail so that you can easily adapt it for your own projects, but feel free to reach out with any questions! -Lacey</p>


</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/lwheinsberg\.github\.io\/mvNUR");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./bnlearn.html" class="pagination-link  aria-label=" &lt;span="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">bnlearn</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Troubleshooting.html" class="pagination-link" aria-label="<span class='chapter-number'>A</span>&nbsp; <span class='chapter-title'>Troubleshooting mvBIMBAM installation</span>">
        <span class="nav-page-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Troubleshooting mvBIMBAM installation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>