<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>A Guide to Multivariate Bayesian Analyses in Nursing Research - Appendix A — Define bnlearn functions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./mvbimbam.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./bnlearnfunctions.html">Appendices</a></li><li class="breadcrumb-item"><a href="./bnlearnfunctions.html"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Define bnlearn functions</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">A Guide to Multivariate Bayesian Analyses in Nursing Research</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prep.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bnlearn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">bnlearn</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mvbimbam.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">mvBIMBAM</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bnlearnfunctions.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Define bnlearn functions</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#function-1-fit.the.model" id="toc-function-1-fit.the.model" class="nav-link active" data-scroll-target="#function-1-fit.the.model"><span class="header-section-number">A.1</span> FUNCTION 1: fit.the.model</a></li>
  <li><a href="#the-bayesian-network-model-is-constructed-using-the-fit.the.model-function" id="toc-the-bayesian-network-model-is-constructed-using-the-fit.the.model-function" class="nav-link" data-scroll-target="#the-bayesian-network-model-is-constructed-using-the-fit.the.model-function"><span class="header-section-number">B</span> The Bayesian network model is constructed using the <code>fit.the.model()</code> function,</a></li>
  <li><a href="#which-employs-a-constrained-learning-algorithm-based-on-conditional-independence" id="toc-which-employs-a-constrained-learning-algorithm-based-on-conditional-independence" class="nav-link" data-scroll-target="#which-employs-a-constrained-learning-algorithm-based-on-conditional-independence"><span class="header-section-number">C</span> which employs a constrained learning algorithm based on conditional independence</a></li>
  <li><a href="#testing-specifically-the-semi-interleaved-hiton-pc-method.-in-this-process-the" id="toc-testing-specifically-the-semi-interleaved-hiton-pc-method.-in-this-process-the" class="nav-link" data-scroll-target="#testing-specifically-the-semi-interleaved-hiton-pc-method.-in-this-process-the"><span class="header-section-number">D</span> testing, specifically the semi-interleaved HITON-PC method. In this process, the</a></li>
  <li><a href="#algorithm-identifies-parent-child-relationships-within-the-network-where-nodes" id="toc-algorithm-identifies-parent-child-relationships-within-the-network-where-nodes" class="nav-link" data-scroll-target="#algorithm-identifies-parent-child-relationships-within-the-network-where-nodes"><span class="header-section-number">E</span> algorithm identifies “parent-child” relationships within the network, where nodes</a></li>
  <li><a href="#represent-either-phenotypic-traits-or-genetic-variants-and-edges-signify-associations" id="toc-represent-either-phenotypic-traits-or-genetic-variants-and-edges-signify-associations" class="nav-link" data-scroll-target="#represent-either-phenotypic-traits-or-genetic-variants-and-edges-signify-associations"><span class="header-section-number">F</span> represent either phenotypic traits or genetic variants, and edges signify associations</a></li>
  <li><a href="#between-these-nodes.-notably-traits-can-have-parent-nodes-that-are-either-genetic" id="toc-between-these-nodes.-notably-traits-can-have-parent-nodes-that-are-either-genetic" class="nav-link" data-scroll-target="#between-these-nodes.-notably-traits-can-have-parent-nodes-that-are-either-genetic"><span class="header-section-number">G</span> between these nodes. Notably, traits can have parent nodes that are either genetic</a></li>
  <li><a href="#variants-or-other-traits-but-they-can-only-serve-as-parents-to-other-traits-adhering" id="toc-variants-or-other-traits-but-they-can-only-serve-as-parents-to-other-traits-adhering" class="nav-link" data-scroll-target="#variants-or-other-traits-but-they-can-only-serve-as-parents-to-other-traits-adhering"><span class="header-section-number">H</span> variants or other traits, but they can only serve as parents to other traits, adhering</a></li>
  <li><a href="#to-the-constraint-that-genetic-variants-do-not-act-on-traits-since-genotypes-are-constant" id="toc-to-the-constraint-that-genetic-variants-do-not-act-on-traits-since-genotypes-are-constant" class="nav-link" data-scroll-target="#to-the-constraint-that-genetic-variants-do-not-act-on-traits-since-genotypes-are-constant"><span class="header-section-number">I</span> to the constraint that genetic variants do not act on traits (since genotypes are constant</a></li>
  <li><a href="#across-an-individuals-lifespan." id="toc-across-an-individuals-lifespan." class="nav-link" data-scroll-target="#across-an-individuals-lifespan."><span class="header-section-number">J</span> across an individual’s lifespan).</a></li>
  <li><a href="#to-enforce-this-constraint-a-blacklist-is-created-using-the-tiers2blacklist" id="toc-to-enforce-this-constraint-a-blacklist-is-created-using-the-tiers2blacklist" class="nav-link" data-scroll-target="#to-enforce-this-constraint-a-blacklist-is-created-using-the-tiers2blacklist"><span class="header-section-number">K</span> To enforce this constraint, a “blacklist” is created using the <code>tiers2blacklist()</code></a></li>
  <li><a href="#function-preventing-arcs-from-being-directed-towards-specific-nodes.-this-restriction" id="toc-function-preventing-arcs-from-being-directed-towards-specific-nodes.-this-restriction" class="nav-link" data-scroll-target="#function-preventing-arcs-from-being-directed-towards-specific-nodes.-this-restriction"><span class="header-section-number">L</span> function, preventing arcs from being directed towards specific nodes. This restriction</a></li>
  <li><a href="#aims-to-guide-the-learning-process-by-ensuring-that-known-causal-relationships-are" id="toc-aims-to-guide-the-learning-process-by-ensuring-that-known-causal-relationships-are" class="nav-link" data-scroll-target="#aims-to-guide-the-learning-process-by-ensuring-that-known-causal-relationships-are"><span class="header-section-number">M</span> aims to guide the learning process by ensuring that known causal relationships are</a></li>
  <li><a href="#inferred-in-the-correct-direction-from-genetic-variants-to-traits-while-also-allowing" id="toc-inferred-in-the-correct-direction-from-genetic-variants-to-traits-while-also-allowing" class="nav-link" data-scroll-target="#inferred-in-the-correct-direction-from-genetic-variants-to-traits-while-also-allowing"><span class="header-section-number">N</span> inferred in the correct direction from genetic variants to traits, while also allowing</a></li>
  <li><a href="#for-customization-to-blacklist-other-traits-if-needed-e.g.-you-can-customize-this" id="toc-for-customization-to-blacklist-other-traits-if-needed-e.g.-you-can-customize-this" class="nav-link" data-scroll-target="#for-customization-to-blacklist-other-traits-if-needed-e.g.-you-can-customize-this"><span class="header-section-number">O</span> for customization to blacklist other traits if needed (e.g., you can customize this</a></li>
  <li><a href="#to-forcerestrict-a-trait-trait-relationship." id="toc-to-forcerestrict-a-trait-trait-relationship." class="nav-link" data-scroll-target="#to-forcerestrict-a-trait-trait-relationship."><span class="header-section-number">P</span> to force/restrict a trait-&gt;trait relationship).</a></li>
  <li><a href="#after-the-networks-are-learned-the-nodes-are-categorized-into-subsets-for" id="toc-after-the-networks-are-learned-the-nodes-are-categorized-into-subsets-for" class="nav-link" data-scroll-target="#after-the-networks-are-learned-the-nodes-are-categorized-into-subsets-for"><span class="header-section-number">Q</span> After the networks are learned, the nodes are categorized into subsets for</a></li>
  <li><a href="#visualization-and-the-network-structures-are-determined-by-maximizing-the-bayesian" id="toc-visualization-and-the-network-structures-are-determined-by-maximizing-the-bayesian" class="nav-link" data-scroll-target="#visualization-and-the-network-structures-are-determined-by-maximizing-the-bayesian"><span class="header-section-number">R</span> visualization, and the network structures are determined by maximizing the Bayesian</a></li>
  <li><a href="#information-criteria-bic.-this-approach-facilitates-the-construction-of-bayesian" id="toc-information-criteria-bic.-this-approach-facilitates-the-construction-of-bayesian" class="nav-link" data-scroll-target="#information-criteria-bic.-this-approach-facilitates-the-construction-of-bayesian"><span class="header-section-number">S</span> Information Criteria (BIC). This approach facilitates the construction of Bayesian</a></li>
  <li><a href="#networks-that-capture-probabilistic-relationships-between-traits-and-genetic-variants" id="toc-networks-that-capture-probabilistic-relationships-between-traits-and-genetic-variants" class="nav-link" data-scroll-target="#networks-that-capture-probabilistic-relationships-between-traits-and-genetic-variants"><span class="header-section-number">T</span> networks that capture probabilistic relationships between traits and genetic variants,</a></li>
  <li><a href="#with-the-learned-structures-reflecting-potential-causal-associations." id="toc-with-the-learned-structures-reflecting-potential-causal-associations." class="nav-link" data-scroll-target="#with-the-learned-structures-reflecting-potential-causal-associations."><span class="header-section-number">U</span> with the learned structures reflecting potential causal associations.</a></li>
  <li><a href="#further-details-on-the-methodology-can-be-found-at" id="toc-further-details-on-the-methodology-can-be-found-at" class="nav-link" data-scroll-target="#further-details-on-the-methodology-can-be-found-at"><span class="header-section-number">V</span> Further details on the methodology can be found at</a></li>
  <li><a href="#httpwww.bnlearn.comresearchgenetics14." id="toc-httpwww.bnlearn.comresearchgenetics14." class="nav-link" data-scroll-target="#httpwww.bnlearn.comresearchgenetics14."><span class="header-section-number">W</span> http://www.bnlearn.com/research/genetics14/.</a></li>
  <li><a href="#data-the-data-set-to-be-used" id="toc-data-the-data-set-to-be-used" class="nav-link" data-scroll-target="#data-the-data-set-to-be-used"><span class="header-section-number">X</span> data: The data set to be used</a></li>
  <li><a href="#alpha-the-significance-level-for-conditional-independence-tests" id="toc-alpha-the-significance-level-for-conditional-independence-tests" class="nav-link" data-scroll-target="#alpha-the-significance-level-for-conditional-independence-tests"><span class="header-section-number">Y</span> alpha: The significance level for conditional independence tests</a>
  <ul class="collapse">
  <li><a href="#function-2-xval.the.model" id="toc-function-2-xval.the.model" class="nav-link" data-scroll-target="#function-2-xval.the.model"><span class="header-section-number">Y.1</span> FUNCTION 2: xval.the.model</a></li>
  </ul></li>
  <li><a href="#the-xval.the.model-function-performs-model-training-with-n-fold" id="toc-the-xval.the.model-function-performs-model-training-with-n-fold" class="nav-link" data-scroll-target="#the-xval.the.model-function-performs-model-training-with-n-fold"><span class="header-section-number">Z</span> The <code>xval.the.model()</code> function performs model training with n-fold</a></li>
  <li><a href="#cross-validation-in-the-case-of-this-example-5-fold-cross-validation." id="toc-cross-validation-in-the-case-of-this-example-5-fold-cross-validation." class="nav-link" data-scroll-target="#cross-validation-in-the-case-of-this-example-5-fold-cross-validation."><span class="header-section-number">[</span> cross-validation (in the case of this example, 5-fold cross-validation).</a></li>
  <li><a href="#during-this-process-the-data-set-is-divided-into-multiple-partitions" id="toc-during-this-process-the-data-set-is-divided-into-multiple-partitions" class="nav-link" data-scroll-target="#during-this-process-the-data-set-is-divided-into-multiple-partitions"><span class="header-section-number">\</span> During this process, the data set is divided into multiple partitions,</a></li>
  <li><a href="#with-each-partition-serving-as-the-test-set-while-the-remaining-data-is-used" id="toc-with-each-partition-serving-as-the-test-set-while-the-remaining-data-is-used" class="nav-link" data-scroll-target="#with-each-partition-serving-as-the-test-set-while-the-remaining-data-is-used"><span class="header-section-number">]</span> with each partition serving as the test set while the remaining data is used</a></li>
  <li><a href="#for-training.-this-process-is-repeated-iteratively-to-ensure-that-all" id="toc-for-training.-this-process-is-repeated-iteratively-to-ensure-that-all" class="nav-link" data-scroll-target="#for-training.-this-process-is-repeated-iteratively-to-ensure-that-all"><span class="header-section-number">^</span> for training. This process is repeated iteratively to ensure that all</a></li>
  <li><a href="#data-points-are-included-in-the-test-set-at-least-once." id="toc-data-points-are-included-in-the-test-set-at-least-once." class="nav-link" data-scroll-target="#data-points-are-included-in-the-test-set-at-least-once."><span class="header-section-number">_</span> data points are included in the test set at least once.</a></li>
  <li><a href="#during-each-fold-of-cross-validation-the-following-steps-are-performed" id="toc-during-each-fold-of-cross-validation-the-following-steps-are-performed" class="nav-link" data-scroll-target="#during-each-fold-of-cross-validation-the-following-steps-are-performed"><span class="header-section-number">`</span> During each fold of cross-validation, the following steps are performed:</a></li>
  <li><a href="#data-splitting-the-dataset-is-divided-into-a-training-set-dtraining" id="toc-data-splitting-the-dataset-is-divided-into-a-training-set-dtraining" class="nav-link" data-scroll-target="#data-splitting-the-dataset-is-divided-into-a-training-set-dtraining"><span class="header-section-number">a</span> (1) Data Splitting: The dataset is divided into a training set (dtraining)</a></li>
  <li><a href="#and-a-test-set-dtest." id="toc-and-a-test-set-dtest." class="nav-link" data-scroll-target="#and-a-test-set-dtest."><span class="header-section-number">b</span> and a test set (dtest).</a></li>
  <li><a href="#model-fitting-a-bayesian-network-model-is-fitted-to-the-training-data" id="toc-model-fitting-a-bayesian-network-model-is-fitted-to-the-training-data" class="nav-link" data-scroll-target="#model-fitting-a-bayesian-network-model-is-fitted-to-the-training-data"><span class="header-section-number">c</span> (2) Model Fitting: A Bayesian network model is fitted to the training data</a></li>
  <li><a href="#using-the-fit.the.model-function.-this-model-captures-the-probabilistic" id="toc-using-the-fit.the.model-function.-this-model-captures-the-probabilistic" class="nav-link" data-scroll-target="#using-the-fit.the.model-function.-this-model-captures-the-probabilistic"><span class="header-section-number">d</span> using the <code>fit.the.model()</code> function. This model captures the probabilistic</a></li>
  <li><a href="#relationships-between-phenotypic-traits-and-genetic-variants." id="toc-relationships-between-phenotypic-traits-and-genetic-variants." class="nav-link" data-scroll-target="#relationships-between-phenotypic-traits-and-genetic-variants."><span class="header-section-number">e</span> relationships between phenotypic traits and genetic variants.</a></li>
  <li><a href="#prediction-the-model-is-used-to-predict-the-values-of-phenotypic-traits" id="toc-prediction-the-model-is-used-to-predict-the-values-of-phenotypic-traits" class="nav-link" data-scroll-target="#prediction-the-model-is-used-to-predict-the-values-of-phenotypic-traits"><span class="header-section-number">f</span> (3) Prediction: The model is used to predict the values of phenotypic traits</a></li>
  <li><a href="#on-the-test-set.-these-predictions-are-stored-in-the-prediction-matrix." id="toc-on-the-test-set.-these-predictions-are-stored-in-the-prediction-matrix." class="nav-link" data-scroll-target="#on-the-test-set.-these-predictions-are-stored-in-the-prediction-matrix."><span class="header-section-number">g</span> on the test set. These predictions are stored in the prediction matrix.</a></li>
  <li><a href="#posterior-estimation-posterior-estimates-are-computed-for-each-trait-based" id="toc-posterior-estimation-posterior-estimates-are-computed-for-each-trait-based" class="nav-link" data-scroll-target="#posterior-estimation-posterior-estimates-are-computed-for-each-trait-based"><span class="header-section-number">h</span> (4) Posterior Estimation: Posterior estimates are computed for each trait based</a></li>
  <li><a href="#on-the-test-data.-these-estimates-are-stored-in-the-post-matrix." id="toc-on-the-test-data.-these-estimates-are-stored-in-the-post-matrix." class="nav-link" data-scroll-target="#on-the-test-data.-these-estimates-are-stored-in-the-post-matrix."><span class="header-section-number">i</span> on the test data. These estimates are stored in the post matrix.</a></li>
  <li><a href="#correlation-computation-the-correlations-between-the-predicted-and-observed" id="toc-correlation-computation-the-correlations-between-the-predicted-and-observed" class="nav-link" data-scroll-target="#correlation-computation-the-correlations-between-the-predicted-and-observed"><span class="header-section-number">j</span> (5) Correlation Computation: The correlations between the predicted and observed</a></li>
  <li><a href="#values-for-each-trait-are-calculated-both-before-predcor-and-after-postcor" id="toc-values-for-each-trait-are-calculated-both-before-predcor-and-after-postcor" class="nav-link" data-scroll-target="#values-for-each-trait-are-calculated-both-before-predcor-and-after-postcor"><span class="header-section-number">k</span> values for each trait are calculated, both before (predcor) and after (postcor)</a></li>
  <li><a href="#cross-validation.-these-correlations-provide-a-measure-of-the-models-predictive-performance." id="toc-cross-validation.-these-correlations-provide-a-measure-of-the-models-predictive-performance." class="nav-link" data-scroll-target="#cross-validation.-these-correlations-provide-a-measure-of-the-models-predictive-performance."><span class="header-section-number">l</span> cross-validation. These correlations provide a measure of the model’s predictive performance.</a></li>
  <li><a href="#finally-the-function-returns-various-results-including-the-predicted-values" id="toc-finally-the-function-returns-various-results-including-the-predicted-values" class="nav-link" data-scroll-target="#finally-the-function-returns-various-results-including-the-predicted-values"><span class="header-section-number">m</span> Finally, the function returns various results, including the predicted values,</a></li>
  <li><a href="#posterior-estimates-observed-values-and-correlation-coefficients-for-each-trait." id="toc-posterior-estimates-observed-values-and-correlation-coefficients-for-each-trait." class="nav-link" data-scroll-target="#posterior-estimates-observed-values-and-correlation-coefficients-for-each-trait."><span class="header-section-number">n</span> posterior estimates, observed values, and correlation coefficients for each trait.</a></li>
  <li><a href="#additionally-it-provides-the-learned-models-for-each-fold-of-cross-validation." id="toc-additionally-it-provides-the-learned-models-for-each-fold-of-cross-validation." class="nav-link" data-scroll-target="#additionally-it-provides-the-learned-models-for-each-fold-of-cross-validation."><span class="header-section-number">o</span> Additionally, it provides the learned models for each fold of cross-validation.</a></li>
  <li><a href="#this-process-allows-for-the-assessment-of-how-well-the-bayesian-network-model" id="toc-this-process-allows-for-the-assessment-of-how-well-the-bayesian-network-model" class="nav-link" data-scroll-target="#this-process-allows-for-the-assessment-of-how-well-the-bayesian-network-model"><span class="header-section-number">p</span> This process allows for the assessment of how well the Bayesian network model</a></li>
  <li><a href="#generalizes-to-unseen-data-and-provides-insights-into-its-predictive-capabilities.-end-redraw.label.ggnet" id="toc-generalizes-to-unseen-data-and-provides-insights-into-its-predictive-capabilities.-end-redraw.label.ggnet" class="nav-link" data-scroll-target="#generalizes-to-unseen-data-and-provides-insights-into-its-predictive-capabilities.-end-redraw.label.ggnet"><span class="header-section-number">q</span> generalizes to unseen data and provides insights into its predictive capabilities. } # End redraw.label.ggnet</a></li>
  <li><a href="#data-the-data-set-to-be-used-1" id="toc-data-the-data-set-to-be-used-1" class="nav-link" data-scroll-target="#data-the-data-set-to-be-used-1"><span class="header-section-number">r</span> data: The data set to be used</a></li>
  <li><a href="#k_crossval-the-number-of-folds-for-cross-validation" id="toc-k_crossval-the-number-of-folds-for-cross-validation" class="nav-link" data-scroll-target="#k_crossval-the-number-of-folds-for-cross-validation"><span class="header-section-number">s</span> k_crossval: The number of folds for cross-validation</a></li>
  <li><a href="#cluster-the-number-of-clusters-to-be-used-for-parallel-processing" id="toc-cluster-the-number-of-clusters-to-be-used-for-parallel-processing" class="nav-link" data-scroll-target="#cluster-the-number-of-clusters-to-be-used-for-parallel-processing"><span class="header-section-number">t</span> cluster: The number of clusters to be used for parallel processing</a></li>
  <li><a href="#alpha-the-significance-level-for-conditional-independence-tests-1" id="toc-alpha-the-significance-level-for-conditional-independence-tests-1" class="nav-link" data-scroll-target="#alpha-the-significance-level-for-conditional-independence-tests-1"><span class="header-section-number">u</span> alpha: The significance level for conditional independence tests</a></li>
  <li><a href="#ridge-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" id="toc-ridge-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" class="nav-link" data-scroll-target="#ridge-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization"><span class="header-section-number">v</span> ridge: A logical value indicating whether to use a custom threshold for network visualization</a></li>
  <li><a href="#create-a-function-to-train-the-model-with-n-fold-cross-validation" id="toc-create-a-function-to-train-the-model-with-n-fold-cross-validation" class="nav-link" data-scroll-target="#create-a-function-to-train-the-model-with-n-fold-cross-validation"><span class="header-section-number">w</span> Create a function to train the model with n fold cross validation,</a></li>
  <li><a href="#n-train-and-100n-test" id="toc-n-train-and-100n-test" class="nav-link" data-scroll-target="#n-train-and-100n-test"><span class="header-section-number">x</span> 100-100/n% train and 100/n% test</a></li>
  <li><a href="#hold-out-each-partition-iteratively-so-that-no-data-are-totally-held-out" id="toc-hold-out-each-partition-iteratively-so-that-no-data-are-totally-held-out" class="nav-link" data-scroll-target="#hold-out-each-partition-iteratively-so-that-no-data-are-totally-held-out"><span class="header-section-number">y</span> Hold out each partition iteratively so that no data are totally held out</a>
  <ul class="collapse">
  <li><a href="#function-3-run_plot_graph" id="toc-function-3-run_plot_graph" class="nav-link" data-scroll-target="#function-3-run_plot_graph"><span class="header-section-number">y.1</span> FUNCTION 3: run_plot_graph</a></li>
  </ul></li>
  <li><a href="#this-code-defines-an-r-function-called-run_plot_graph-that-calling-the-functions" id="toc-this-code-defines-an-r-function-called-run_plot_graph-that-calling-the-functions" class="nav-link" data-scroll-target="#this-code-defines-an-r-function-called-run_plot_graph-that-calling-the-functions"><span class="header-section-number">z</span> This code defines an R function called <code>run_plot_graph</code> that, calling the functions</a></li>
  <li><a href="#above-learns-and-visualizes-the-bayesian-networks-in-the-data-including-customizing" id="toc-above-learns-and-visualizes-the-bayesian-networks-in-the-data-including-customizing" class="nav-link" data-scroll-target="#above-learns-and-visualizes-the-bayesian-networks-in-the-data-including-customizing"><span class="header-section-number">{</span> above, learns and visualizes the Bayesian networks in the data, including customizing</a></li>
  <li><a href="#node-labels-and-appearance-and-then-returns-results-related-to-the-network-analysis." id="toc-node-labels-and-appearance-and-then-returns-results-related-to-the-network-analysis." class="nav-link" data-scroll-target="#node-labels-and-appearance-and-then-returns-results-related-to-the-network-analysis."><span class="header-section-number">|</span> node labels and appearance, and then returns results related to the network analysis.</a></li>
  <li><a href="#note-that-the-following-function-uses-parallel-computing-which-involves-breaking-down" id="toc-note-that-the-following-function-uses-parallel-computing-which-involves-breaking-down" class="nav-link" data-scroll-target="#note-that-the-following-function-uses-parallel-computing-which-involves-breaking-down"><span class="header-section-number">}</span> Note that the following function uses parallel computing which involves breaking down</a></li>
  <li><a href="#a-complex-task-into-smaller-subtasks-that-can-be-executed-simultaneously" id="toc-a-complex-task-into-smaller-subtasks-that-can-be-executed-simultaneously" class="nav-link" data-scroll-target="#a-complex-task-into-smaller-subtasks-that-can-be-executed-simultaneously"><span class="header-section-number">~</span> a complex task into smaller subtasks that can be executed simultaneously,</a></li>
  <li><a href="#or-in-parallel-by-multiple-processors-or-computers-to-speed-up-computationally" id="toc-or-in-parallel-by-multiple-processors-or-computers-to-speed-up-computationally" class="nav-link" data-scroll-target="#or-in-parallel-by-multiple-processors-or-computers-to-speed-up-computationally"><span class="header-section-number"></span> or in parallel, by multiple processors or computers to speed up computationally</a></li>
  <li><a href="#intense-calculations." id="toc-intense-calculations." class="nav-link" data-scroll-target="#intense-calculations."><span class="header-section-number">�</span> intense calculations.</a></li>
  <li><a href="#data-the-dataset-to-be-used-for-plotting-the-networks" id="toc-data-the-dataset-to-be-used-for-plotting-the-networks" class="nav-link" data-scroll-target="#data-the-dataset-to-be-used-for-plotting-the-networks"><span class="header-section-number">�</span> data: The dataset to be used for plotting the networks</a></li>
  <li><a href="#k_crossval-the-number-of-folds-for-cross-validation-1" id="toc-k_crossval-the-number-of-folds-for-cross-validation-1" class="nav-link" data-scroll-target="#k_crossval-the-number-of-folds-for-cross-validation-1"><span class="header-section-number">�</span> k_crossval: The number of folds for cross-validation</a></li>
  <li><a href="#k_iterations-the-number-of-times-the-entire-process-which-includes-k-fold-cross-validation-is-repeated." id="toc-k_iterations-the-number-of-times-the-entire-process-which-includes-k-fold-cross-validation-is-repeated." class="nav-link" data-scroll-target="#k_iterations-the-number-of-times-the-entire-process-which-includes-k-fold-cross-validation-is-repeated."><span class="header-section-number">�</span> k_iterations: The number of times the entire process (which includes k-fold cross-validation) is repeated.</a></li>
  <li><a href="#alpha-the-significance-level-for-conditional-independence-tests-2" id="toc-alpha-the-significance-level-for-conditional-independence-tests-2" class="nav-link" data-scroll-target="#alpha-the-significance-level-for-conditional-independence-tests-2"><span class="header-section-number">�</span> alpha: The significance level for conditional independence tests</a></li>
  <li><a href="#use.custom.threshold-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" id="toc-use.custom.threshold-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" class="nav-link" data-scroll-target="#use.custom.threshold-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization"><span class="header-section-number">�</span> use.custom.threshold: A logical value indicating whether to use a custom threshold for network visualization</a></li>
  <li><a href="#custom.threshold-the-custom-threshold-value-to-be-used-if-use.custom.threshold-is-set-to-true" id="toc-custom.threshold-the-custom-threshold-value-to-be-used-if-use.custom.threshold-is-set-to-true" class="nav-link" data-scroll-target="#custom.threshold-the-custom-threshold-value-to-be-used-if-use.custom.threshold-is-set-to-true"><span class="header-section-number">�</span> custom.threshold: The custom threshold value to be used if use.custom.threshold is set to TRUE</a></li>
  <li><a href="#ncluster-the-number-of-clusters-for-parallel-computation" id="toc-ncluster-the-number-of-clusters-for-parallel-computation" class="nav-link" data-scroll-target="#ncluster-the-number-of-clusters-for-parallel-computation"><span class="header-section-number">�</span> ncluster: The number of clusters for parallel computation</a></li>
  <li><a href="#additional-arguments-to-be-passed-to-other-functions" id="toc-additional-arguments-to-be-passed-to-other-functions" class="nav-link" data-scroll-target="#additional-arguments-to-be-passed-to-other-functions"><span class="header-section-number">�</span> …: Additional arguments to be passed to other functions</a></li>
  <li><a href="#define-function-to-plot-the-networks" id="toc-define-function-to-plot-the-networks" class="nav-link" data-scroll-target="#define-function-to-plot-the-networks"><span class="header-section-number">�</span> Define function to plot the networks</a>
  <ul class="collapse">
  <li><a href="#function-4-match.arcs.and.directions" id="toc-function-4-match.arcs.and.directions" class="nav-link" data-scroll-target="#function-4-match.arcs.and.directions"><span class="header-section-number">�.1</span> FUNCTION 4: match.arcs.and.directions</a></li>
  </ul></li>
  <li><a href="#this-code-defines-an-r-function-called-match.arcs.and.directions-which-is" id="toc-this-code-defines-an-r-function-called-match.arcs.and.directions-which-is" class="nav-link" data-scroll-target="#this-code-defines-an-r-function-called-match.arcs.and.directions-which-is"><span class="header-section-number">�</span> This code defines an R function called <code>match.arcs.and.directions</code> which is</a></li>
  <li><a href="#designed-to-match-arcs-in-a-network-with-their-corresponding-direction" id="toc-designed-to-match-arcs-in-a-network-with-their-corresponding-direction" class="nav-link" data-scroll-target="#designed-to-match-arcs-in-a-network-with-their-corresponding-direction"><span class="header-section-number">�</span> designed to match arcs in a network with their corresponding direction</a></li>
  <li><a href="#coefficients-in-a-strengths-data-frame-providing-flexibility-to-either" id="toc-coefficients-in-a-strengths-data-frame-providing-flexibility-to-either" class="nav-link" data-scroll-target="#coefficients-in-a-strengths-data-frame-providing-flexibility-to-either"><span class="header-section-number">�</span> coefficients in a strengths data frame, providing flexibility to either</a></li>
  <li><a href="#keep-or-discard-additional-information-based-on-the-keep-argument." id="toc-keep-or-discard-additional-information-based-on-the-keep-argument." class="nav-link" data-scroll-target="#keep-or-discard-additional-information-based-on-the-keep-argument."><span class="header-section-number">�</span> keep or discard additional information based on the keep argument.</a></li>
  <li><a href="#arcs-arcs-within-a-bayesian-network-representing-relationships" id="toc-arcs-arcs-within-a-bayesian-network-representing-relationships" class="nav-link" data-scroll-target="#arcs-arcs-within-a-bayesian-network-representing-relationships"><span class="header-section-number">�</span> arcs: Arcs within a Bayesian network representing relationships</a></li>
  <li><a href="#between-nodes-genes-or-traits" id="toc-between-nodes-genes-or-traits" class="nav-link" data-scroll-target="#between-nodes-genes-or-traits"><span class="header-section-number">�</span> between nodes (genes or traits)</a></li>
  <li><a href="#nodes-nodes-within-a-bayesian-network-which-can-represent" id="toc-nodes-nodes-within-a-bayesian-network-which-can-represent" class="nav-link" data-scroll-target="#nodes-nodes-within-a-bayesian-network-which-can-represent"><span class="header-section-number">�</span> nodes: Nodes within a Bayesian network, which can represent</a></li>
  <li><a href="#either-genes-or-traits" id="toc-either-genes-or-traits" class="nav-link" data-scroll-target="#either-genes-or-traits"><span class="header-section-number">�</span> either genes or traits</a></li>
  <li><a href="#strengths-strength-coefficients-associated-with-the-arcs-in" id="toc-strengths-strength-coefficients-associated-with-the-arcs-in" class="nav-link" data-scroll-target="#strengths-strength-coefficients-associated-with-the-arcs-in"><span class="header-section-number">�</span> strengths: Strength coefficients associated with the arcs in</a></li>
  <li><a href="#the-bayesian-network-indicating-the-strength-and-direction" id="toc-the-bayesian-network-indicating-the-strength-and-direction" class="nav-link" data-scroll-target="#the-bayesian-network-indicating-the-strength-and-direction"><span class="header-section-number">�</span> the Bayesian network, indicating the strength and direction</a></li>
  <li><a href="#of-relationships" id="toc-of-relationships" class="nav-link" data-scroll-target="#of-relationships"><span class="header-section-number">�</span> of relationships</a></li>
  <li><a href="#keep-a-logical-value-indicating-whether-to-keep-additional" id="toc-keep-a-logical-value-indicating-whether-to-keep-additional" class="nav-link" data-scroll-target="#keep-a-logical-value-indicating-whether-to-keep-additional"><span class="header-section-number">�</span> keep: A logical value indicating whether to keep additional</a></li>
  <li><a href="#information-when-matching-arcs-and-directions-in-the" id="toc-information-when-matching-arcs-and-directions-in-the" class="nav-link" data-scroll-target="#information-when-matching-arcs-and-directions-in-the"><span class="header-section-number">�</span> information when matching arcs and directions in the</a></li>
  <li><a href="#bayesian-network" id="toc-bayesian-network" class="nav-link" data-scroll-target="#bayesian-network"><span class="header-section-number">�</span> Bayesian network</a></li>
  <li><a href="#define-function-to-match-arcs-and-directions" id="toc-define-function-to-match-arcs-and-directions" class="nav-link" data-scroll-target="#define-function-to-match-arcs-and-directions"><span class="header-section-number">�</span> Define function to match arcs and directions</a>
  <ul class="collapse">
  <li><a href="#function-5-redraw.graph.labels" id="toc-function-5-redraw.graph.labels" class="nav-link" data-scroll-target="#function-5-redraw.graph.labels"><span class="header-section-number">�.1</span> FUNCTION 5: redraw.graph.labels</a></li>
  </ul></li>
  <li><a href="#this-code-defines-the-redraw.graph.labels-function-which-enhances-the-graph" id="toc-this-code-defines-the-redraw.graph.labels-function-which-enhances-the-graph" class="nav-link" data-scroll-target="#this-code-defines-the-redraw.graph.labels-function-which-enhances-the-graph"><span class="header-section-number">�</span> This code defines the <code>redraw.graph.labels</code> function, which enhances the graph</a></li>
  <li><a href="#created-above-by-adding-labels-to-its-edges.-these-labels-make-the-graph-more" id="toc-created-above-by-adding-labels-to-its-edges.-these-labels-make-the-graph-more" class="nav-link" data-scroll-target="#created-above-by-adding-labels-to-its-edges.-these-labels-make-the-graph-more"><span class="header-section-number">�</span> created above by adding labels to its edges. These labels make the graph more</a></li>
  <li><a href="#informative-by-conveying-information-about-the-strength-and-direction-of" id="toc-informative-by-conveying-information-about-the-strength-and-direction-of" class="nav-link" data-scroll-target="#informative-by-conveying-information-about-the-strength-and-direction-of"><span class="header-section-number">�</span> informative by conveying information about the strength and direction of</a></li>
  <li><a href="#relationships-represented-by-each-edge-in-the-graph." id="toc-relationships-represented-by-each-edge-in-the-graph." class="nav-link" data-scroll-target="#relationships-represented-by-each-edge-in-the-graph."><span class="header-section-number">�</span> relationships represented by each edge in the graph.</a></li>
  <li><a href="#strength-a-measure-of-confidence-of-that-edge-while-fixing-the-rest-of" id="toc-strength-a-measure-of-confidence-of-that-edge-while-fixing-the-rest-of" class="nav-link" data-scroll-target="#strength-a-measure-of-confidence-of-that-edge-while-fixing-the-rest-of"><span class="header-section-number">�</span> Strength: a measure of confidence of that edge while fixing the rest of</a></li>
  <li><a href="#the-network-structure-and-is-defined-as-the-empirical-frequency-a-specific" id="toc-the-network-structure-and-is-defined-as-the-empirical-frequency-a-specific" class="nav-link" data-scroll-target="#the-network-structure-and-is-defined-as-the-empirical-frequency-a-specific"><span class="header-section-number">�</span> the network structure and is defined as the empirical frequency a specific</a></li>
  <li><a href="#edge-is-observed-over-a-set-of-networks-learned-from-iterations" id="toc-edge-is-observed-over-a-set-of-networks-learned-from-iterations" class="nav-link" data-scroll-target="#edge-is-observed-over-a-set-of-networks-learned-from-iterations"><span class="header-section-number">�</span> edge is observed over a set of networks learned from iterations</a></li>
  <li><a href="#i.e.-the-number-of-times-the-edge-was-present-out-of-the-total-number-of" id="toc-i.e.-the-number-of-times-the-edge-was-present-out-of-the-total-number-of" class="nav-link" data-scroll-target="#i.e.-the-number-of-times-the-edge-was-present-out-of-the-total-number-of"><span class="header-section-number">�</span> (i.e., the number of times the edge was present out of the total number of</a></li>
  <li><a href="#iterations." id="toc-iterations." class="nav-link" data-scroll-target="#iterations."><span class="header-section-number">�</span> iterations).</a></li>
  <li><a href="#direction-the-probability-of-the-edges-direction-conditional-on-the-edges" id="toc-direction-the-probability-of-the-edges-direction-conditional-on-the-edges" class="nav-link" data-scroll-target="#direction-the-probability-of-the-edges-direction-conditional-on-the-edges"><span class="header-section-number">�</span> Direction: the probability of the edge’s direction conditional on the edge’s</a></li>
  <li><a href="#presence-within-the-network-i.e.-the-number-of-times-the-edge-traveled-in" id="toc-presence-within-the-network-i.e.-the-number-of-times-the-edge-traveled-in" class="nav-link" data-scroll-target="#presence-within-the-network-i.e.-the-number-of-times-the-edge-traveled-in"><span class="header-section-number">�</span> presence within the network (i.e., the number of times the edge traveled in</a></li>
  <li><a href="#a-specific-direction-out-of-the-total-number-of-iterations" id="toc-a-specific-direction-out-of-the-total-number-of-iterations" class="nav-link" data-scroll-target="#a-specific-direction-out-of-the-total-number-of-iterations"><span class="header-section-number">�</span> a specific direction out of the total number of iterations</a></li>
  <li><a href="#in-which-it-was-present." id="toc-in-which-it-was-present." class="nav-link" data-scroll-target="#in-which-it-was-present."><span class="header-section-number">�</span> in which it was present).</a></li>
  <li><a href="#averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only" id="toc-averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only" class="nav-link" data-scroll-target="#averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only"><span class="header-section-number">�</span> averaged2: Bayesian network learned through functions above (relevant nodes only)</a></li>
  <li><a href="#strength2-bayesian-network-strengths-learned-through-the-functions-above" id="toc-strength2-bayesian-network-strengths-learned-through-the-functions-above" class="nav-link" data-scroll-target="#strength2-bayesian-network-strengths-learned-through-the-functions-above"><span class="header-section-number">�</span> strength2: Bayesian network strengths learned through the functions above</a></li>
  <li><a href="#averaged-bayesian-network-learned-through-functions-above-all-nodes" id="toc-averaged-bayesian-network-learned-through-functions-above-all-nodes" class="nav-link" data-scroll-target="#averaged-bayesian-network-learned-through-functions-above-all-nodes"><span class="header-section-number">�</span> averaged: Bayesian network learned through functions above (all nodes)</a></li>
  <li><a href="#traits-phenotypes-of-interest" id="toc-traits-phenotypes-of-interest" class="nav-link" data-scroll-target="#traits-phenotypes-of-interest"><span class="header-section-number">�</span> traits: Phenotypes of interest</a></li>
  <li><a href="#custom.threshold-custom-threshold-to-customize-line-thicknesstype" id="toc-custom.threshold-custom-threshold-to-customize-line-thicknesstype" class="nav-link" data-scroll-target="#custom.threshold-custom-threshold-to-customize-line-thicknesstype"><span class="header-section-number">�</span> custom.threshold: Custom threshold to customize line thickness/type</a></li>
  <li><a href="#define-function-to-redraw-graph-labels" id="toc-define-function-to-redraw-graph-labels" class="nav-link" data-scroll-target="#define-function-to-redraw-graph-labels"><span class="header-section-number">�</span> Define function to redraw graph labels</a>
  <ul class="collapse">
  <li><a href="#function-6-redraw.label.ggnet" id="toc-function-6-redraw.label.ggnet" class="nav-link" data-scroll-target="#function-6-redraw.label.ggnet"><span class="header-section-number">�.1</span> FUNCTION 6: redraw.label.ggnet</a></li>
  </ul></li>
  <li><a href="#the-following-function-redraw.label.ggnet-is-used-for-an-additional" id="toc-the-following-function-redraw.label.ggnet-is-used-for-an-additional" class="nav-link" data-scroll-target="#the-following-function-redraw.label.ggnet-is-used-for-an-additional"><span class="header-section-number">�</span> The following function, <code>redraw.label.ggnet</code>, is used for an additional</a></li>
  <li><a href="#redrawing-of-the-bayesian-network-but-in-a-ggnetworkggplot2-framework" id="toc-redrawing-of-the-bayesian-network-but-in-a-ggnetworkggplot2-framework" class="nav-link" data-scroll-target="#redrawing-of-the-bayesian-network-but-in-a-ggnetworkggplot2-framework"><span class="header-section-number">�</span> redrawing of the Bayesian network but in a ggnetwork/ggplot2 framework,</a></li>
  <li><a href="#enhancing-it-with-labels-and-color-coding-for-interpretation-in-the-context" id="toc-enhancing-it-with-labels-and-color-coding-for-interpretation-in-the-context" class="nav-link" data-scroll-target="#enhancing-it-with-labels-and-color-coding-for-interpretation-in-the-context"><span class="header-section-number">�</span> enhancing it with labels and color-coding for interpretation in the context</a></li>
  <li><a href="#of-bayesian-analysis.-of-note-this-function-calls-in-ggrepel-so" id="toc-of-bayesian-analysis.-of-note-this-function-calls-in-ggrepel-so" class="nav-link" data-scroll-target="#of-bayesian-analysis.-of-note-this-function-calls-in-ggrepel-so"><span class="header-section-number">�</span> of Bayesian analysis. Of note, this function calls in <code>ggrepel()</code> so</a></li>
  <li><a href="#that-node-and-edge-labels-do-not-overlap-further-supporting-the-user" id="toc-that-node-and-edge-labels-do-not-overlap-further-supporting-the-user" class="nav-link" data-scroll-target="#that-node-and-edge-labels-do-not-overlap-further-supporting-the-user"><span class="header-section-number">�</span> that node and edge labels do not overlap, further supporting the user</a></li>
  <li><a href="#in-creating-publication-quality-figures." id="toc-in-creating-publication-quality-figures." class="nav-link" data-scroll-target="#in-creating-publication-quality-figures."><span class="header-section-number">�</span> in creating publication quality figures.</a></li>
  <li><a href="#averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only-1" id="toc-averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only-1" class="nav-link" data-scroll-target="#averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only-1"><span class="header-section-number">�</span> averaged2: Bayesian network learned through functions above (relevant nodes only)</a></li>
  <li><a href="#strength2-bayesian-network-strengths-learned-through-the-functions-above-1" id="toc-strength2-bayesian-network-strengths-learned-through-the-functions-above-1" class="nav-link" data-scroll-target="#strength2-bayesian-network-strengths-learned-through-the-functions-above-1"><span class="header-section-number">�</span> strength2: Bayesian network strengths learned through the functions above</a></li>
  <li><a href="#averaged-bayesian-network-learned-through-functions-above-all-nodes-1" id="toc-averaged-bayesian-network-learned-through-functions-above-all-nodes-1" class="nav-link" data-scroll-target="#averaged-bayesian-network-learned-through-functions-above-all-nodes-1"><span class="header-section-number">�</span> averaged: Bayesian network learned through functions above (all nodes)</a></li>
  <li><a href="#traits-phenotype-of-interest" id="toc-traits-phenotype-of-interest" class="nav-link" data-scroll-target="#traits-phenotype-of-interest"><span class="header-section-number">�</span> traits: Phenotype of interest</a></li>
  <li><a href="#additional-arguments-to-be-passed-to-other-functions-1" id="toc-additional-arguments-to-be-passed-to-other-functions-1" class="nav-link" data-scroll-target="#additional-arguments-to-be-passed-to-other-functions-1"><span class="header-section-number">�</span> …: Additional arguments to be passed to other functions</a></li>
  <li><a href="#define-function-to-redraw-the-graph-in-ggnetworkggplot2-framework" id="toc-define-function-to-redraw-the-graph-in-ggnetworkggplot2-framework" class="nav-link" data-scroll-target="#define-function-to-redraw-the-graph-in-ggnetworkggplot2-framework"><span class="header-section-number">�</span> Define function to redraw the graph in ggnetwork/ggplot2 framework</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./bnlearnfunctions.html">Appendices</a></li><li class="breadcrumb-item"><a href="./bnlearnfunctions.html"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Define bnlearn functions</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Appendix A — Define bnlearn functions</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="section" class="level51">
<p class="heading"></p>
</section>
<section id="function-1-fit.the.model" class="level2" data-number="A.1">
<h2 data-number="A.1" class="anchored" data-anchor-id="function-1-fit.the.model"><span class="header-section-number">A.1</span> FUNCTION 1: fit.the.model</h2>
</section>
<section id="the-bayesian-network-model-is-constructed-using-the-fit.the.model-function" class="level1" data-number="B">
<h1 data-number="B"><span class="header-section-number">B</span> The Bayesian network model is constructed using the <code>fit.the.model()</code> function,</h1>
</section>
<section id="which-employs-a-constrained-learning-algorithm-based-on-conditional-independence" class="level1" data-number="C">
<h1 data-number="C"><span class="header-section-number">C</span> which employs a constrained learning algorithm based on conditional independence</h1>
</section>
<section id="testing-specifically-the-semi-interleaved-hiton-pc-method.-in-this-process-the" class="level1" data-number="D">
<h1 data-number="D"><span class="header-section-number">D</span> testing, specifically the semi-interleaved HITON-PC method. In this process, the</h1>
</section>
<section id="algorithm-identifies-parent-child-relationships-within-the-network-where-nodes" class="level1" data-number="E">
<h1 data-number="E"><span class="header-section-number">E</span> algorithm identifies “parent-child” relationships within the network, where nodes</h1>
</section>
<section id="represent-either-phenotypic-traits-or-genetic-variants-and-edges-signify-associations" class="level1" data-number="F">
<h1 data-number="F"><span class="header-section-number">F</span> represent either phenotypic traits or genetic variants, and edges signify associations</h1>
</section>
<section id="between-these-nodes.-notably-traits-can-have-parent-nodes-that-are-either-genetic" class="level1" data-number="G">
<h1 data-number="G"><span class="header-section-number">G</span> between these nodes. Notably, traits can have parent nodes that are either genetic</h1>
</section>
<section id="variants-or-other-traits-but-they-can-only-serve-as-parents-to-other-traits-adhering" class="level1" data-number="H">
<h1 data-number="H"><span class="header-section-number">H</span> variants or other traits, but they can only serve as parents to other traits, adhering</h1>
</section>
<section id="to-the-constraint-that-genetic-variants-do-not-act-on-traits-since-genotypes-are-constant" class="level1" data-number="I">
<h1 data-number="I"><span class="header-section-number">I</span> to the constraint that genetic variants do not act on traits (since genotypes are constant</h1>
</section>
<section id="across-an-individuals-lifespan." class="level1" data-number="J">
<h1 data-number="J"><span class="header-section-number">J</span> across an individual’s lifespan).</h1>
</section>
<section id="to-enforce-this-constraint-a-blacklist-is-created-using-the-tiers2blacklist" class="level1" data-number="K">
<h1 data-number="K"><span class="header-section-number">K</span> To enforce this constraint, a “blacklist” is created using the <code>tiers2blacklist()</code></h1>
</section>
<section id="function-preventing-arcs-from-being-directed-towards-specific-nodes.-this-restriction" class="level1" data-number="L">
<h1 data-number="L"><span class="header-section-number">L</span> function, preventing arcs from being directed towards specific nodes. This restriction</h1>
</section>
<section id="aims-to-guide-the-learning-process-by-ensuring-that-known-causal-relationships-are" class="level1" data-number="M">
<h1 data-number="M"><span class="header-section-number">M</span> aims to guide the learning process by ensuring that known causal relationships are</h1>
</section>
<section id="inferred-in-the-correct-direction-from-genetic-variants-to-traits-while-also-allowing" class="level1" data-number="N">
<h1 data-number="N"><span class="header-section-number">N</span> inferred in the correct direction from genetic variants to traits, while also allowing</h1>
</section>
<section id="for-customization-to-blacklist-other-traits-if-needed-e.g.-you-can-customize-this" class="level1" data-number="O">
<h1 data-number="O"><span class="header-section-number">O</span> for customization to blacklist other traits if needed (e.g., you can customize this</h1>
</section>
<section id="to-forcerestrict-a-trait-trait-relationship." class="level1" data-number="P">
<h1 data-number="P"><span class="header-section-number">P</span> to force/restrict a trait-&gt;trait relationship).</h1>
</section>
<section id="after-the-networks-are-learned-the-nodes-are-categorized-into-subsets-for" class="level1" data-number="Q">
<h1 data-number="Q"><span class="header-section-number">Q</span> After the networks are learned, the nodes are categorized into subsets for</h1>
</section>
<section id="visualization-and-the-network-structures-are-determined-by-maximizing-the-bayesian" class="level1" data-number="R">
<h1 data-number="R"><span class="header-section-number">R</span> visualization, and the network structures are determined by maximizing the Bayesian</h1>
</section>
<section id="information-criteria-bic.-this-approach-facilitates-the-construction-of-bayesian" class="level1" data-number="S">
<h1 data-number="S"><span class="header-section-number">S</span> Information Criteria (BIC). This approach facilitates the construction of Bayesian</h1>
</section>
<section id="networks-that-capture-probabilistic-relationships-between-traits-and-genetic-variants" class="level1" data-number="T">
<h1 data-number="T"><span class="header-section-number">T</span> networks that capture probabilistic relationships between traits and genetic variants,</h1>
</section>
<section id="with-the-learned-structures-reflecting-potential-causal-associations." class="level1" data-number="U">
<h1 data-number="U"><span class="header-section-number">U</span> with the learned structures reflecting potential causal associations.</h1>
</section>
<section id="further-details-on-the-methodology-can-be-found-at" class="level1" data-number="V">
<h1 data-number="V"><span class="header-section-number">V</span> Further details on the methodology can be found at</h1>
</section>
<section id="httpwww.bnlearn.comresearchgenetics14." class="level1" data-number="W">
<h1 data-number="W"><span class="header-section-number">W</span> http://www.bnlearn.com/research/genetics14/.</h1>
</section>
<section id="data-the-data-set-to-be-used" class="level1" data-number="X">
<h1 data-number="X"><span class="header-section-number">X</span> data: The data set to be used</h1>
</section>
<section id="alpha-the-significance-level-for-conditional-independence-tests" class="level1" data-number="Y">
<h1 data-number="Y"><span class="header-section-number">Y</span> alpha: The significance level for conditional independence tests</h1>
<p>fit.the.model &lt;- function(data, alpha) { # Start fit.the.model</p>
<p># Creates an empty list (cpc) with a length equal to the length of the traits vector cpc &lt;- vector(length(traits), mode = “list”) # Assign names to the elements of the cpc list based on the values in the traits vector names(cpc) &lt;- traits</p>
<p># Identify the parents of each trait (may be variants or other traits) for (t in seq_along(traits)) { # Start for cpc[[t]] = learn.nbr(data[, c(traits, genes)], node = traits[t], debug = FALSE, method = “si.hiton.pc”, test = “cor”, alpha = alpha) } # End for</p>
<p># Merge the relevant variables to use for learning nodes &lt;- unique(c(traits, unlist(cpc)))</p>
<p># Set up blacklist which specifies that variants cannot depend on traits blacklist &lt;- tiers2blacklist(list(nodes[!(nodes %in% traits)], traits))</p>
<p># Build the Bayesian network bn &lt;- hc(data[, nodes], blacklist = blacklist)</p>
<p># Return Bayesian network return(bn)</p>
<p>} # End fit.the.model</p>
<section id="section-1" class="level51">
<p class="heading"></p>
</section>
<section id="section-2" class="level51">
<p class="heading"></p>
</section>
<section id="function-2-xval.the.model" class="level2" data-number="Y.1">
<h2 data-number="Y.1" class="anchored" data-anchor-id="function-2-xval.the.model"><span class="header-section-number">Y.1</span> FUNCTION 2: xval.the.model</h2>
</section>
</section>
<section id="the-xval.the.model-function-performs-model-training-with-n-fold" class="level1" data-number="Z">
<h1 data-number="Z"><span class="header-section-number">Z</span> The <code>xval.the.model()</code> function performs model training with n-fold</h1>
</section>
<section id="cross-validation-in-the-case-of-this-example-5-fold-cross-validation." class="level1" data-number="[">
<h1 data-number="["><span class="header-section-number">[</span> cross-validation (in the case of this example, 5-fold cross-validation).</h1>
</section>
<section id="during-this-process-the-data-set-is-divided-into-multiple-partitions" class="level1" data-number="\">
<h1 data-number="\"><span class="header-section-number">\</span> During this process, the data set is divided into multiple partitions,</h1>
</section>
<section id="with-each-partition-serving-as-the-test-set-while-the-remaining-data-is-used" class="level1" data-number="]">
<h1 data-number="]"><span class="header-section-number">]</span> with each partition serving as the test set while the remaining data is used</h1>
</section>
<section id="for-training.-this-process-is-repeated-iteratively-to-ensure-that-all" class="level1" data-number="^">
<h1 data-number="^"><span class="header-section-number">^</span> for training. This process is repeated iteratively to ensure that all</h1>
</section>
<section id="data-points-are-included-in-the-test-set-at-least-once." class="level1" data-number="_">
<h1 data-number="_"><span class="header-section-number">_</span> data points are included in the test set at least once.</h1>
</section>
<section id="during-each-fold-of-cross-validation-the-following-steps-are-performed" class="level1" data-number="`">
<h1 data-number="`"><span class="header-section-number">`</span> During each fold of cross-validation, the following steps are performed:</h1>
</section>
<section id="data-splitting-the-dataset-is-divided-into-a-training-set-dtraining" class="level1" data-number="a">
<h1 data-number="a"><span class="header-section-number">a</span> (1) Data Splitting: The dataset is divided into a training set (dtraining)</h1>
</section>
<section id="and-a-test-set-dtest." class="level1" data-number="b">
<h1 data-number="b"><span class="header-section-number">b</span> and a test set (dtest).</h1>
</section>
<section id="model-fitting-a-bayesian-network-model-is-fitted-to-the-training-data" class="level1" data-number="c">
<h1 data-number="c"><span class="header-section-number">c</span> (2) Model Fitting: A Bayesian network model is fitted to the training data</h1>
</section>
<section id="using-the-fit.the.model-function.-this-model-captures-the-probabilistic" class="level1" data-number="d">
<h1 data-number="d"><span class="header-section-number">d</span> using the <code>fit.the.model()</code> function. This model captures the probabilistic</h1>
</section>
<section id="relationships-between-phenotypic-traits-and-genetic-variants." class="level1" data-number="e">
<h1 data-number="e"><span class="header-section-number">e</span> relationships between phenotypic traits and genetic variants.</h1>
</section>
<section id="prediction-the-model-is-used-to-predict-the-values-of-phenotypic-traits" class="level1" data-number="f">
<h1 data-number="f"><span class="header-section-number">f</span> (3) Prediction: The model is used to predict the values of phenotypic traits</h1>
</section>
<section id="on-the-test-set.-these-predictions-are-stored-in-the-prediction-matrix." class="level1" data-number="g">
<h1 data-number="g"><span class="header-section-number">g</span> on the test set. These predictions are stored in the prediction matrix.</h1>
</section>
<section id="posterior-estimation-posterior-estimates-are-computed-for-each-trait-based" class="level1" data-number="h">
<h1 data-number="h"><span class="header-section-number">h</span> (4) Posterior Estimation: Posterior estimates are computed for each trait based</h1>
</section>
<section id="on-the-test-data.-these-estimates-are-stored-in-the-post-matrix." class="level1" data-number="i">
<h1 data-number="i"><span class="header-section-number">i</span> on the test data. These estimates are stored in the post matrix.</h1>
</section>
<section id="correlation-computation-the-correlations-between-the-predicted-and-observed" class="level1" data-number="j">
<h1 data-number="j"><span class="header-section-number">j</span> (5) Correlation Computation: The correlations between the predicted and observed</h1>
</section>
<section id="values-for-each-trait-are-calculated-both-before-predcor-and-after-postcor" class="level1" data-number="k">
<h1 data-number="k"><span class="header-section-number">k</span> values for each trait are calculated, both before (predcor) and after (postcor)</h1>
</section>
<section id="cross-validation.-these-correlations-provide-a-measure-of-the-models-predictive-performance." class="level1" data-number="l">
<h1 data-number="l"><span class="header-section-number">l</span> cross-validation. These correlations provide a measure of the model’s predictive performance.</h1>
</section>
<section id="finally-the-function-returns-various-results-including-the-predicted-values" class="level1" data-number="m">
<h1 data-number="m"><span class="header-section-number">m</span> Finally, the function returns various results, including the predicted values,</h1>
</section>
<section id="posterior-estimates-observed-values-and-correlation-coefficients-for-each-trait." class="level1" data-number="n">
<h1 data-number="n"><span class="header-section-number">n</span> posterior estimates, observed values, and correlation coefficients for each trait.</h1>
</section>
<section id="additionally-it-provides-the-learned-models-for-each-fold-of-cross-validation." class="level1" data-number="o">
<h1 data-number="o"><span class="header-section-number">o</span> Additionally, it provides the learned models for each fold of cross-validation.</h1>
</section>
<section id="this-process-allows-for-the-assessment-of-how-well-the-bayesian-network-model" class="level1" data-number="p">
<h1 data-number="p"><span class="header-section-number">p</span> This process allows for the assessment of how well the Bayesian network model</h1>
</section>
<section id="generalizes-to-unseen-data-and-provides-insights-into-its-predictive-capabilities.-end-redraw.label.ggnet" class="level1" data-number="q">
<h1 data-number="q"><span class="header-section-number">q</span> generalizes to unseen data and provides insights into its predictive capabilities. } # End redraw.label.ggnet</h1>
</section>
<section id="data-the-data-set-to-be-used-1" class="level1" data-number="r">
<h1 data-number="r"><span class="header-section-number">r</span> data: The data set to be used</h1>
</section>
<section id="k_crossval-the-number-of-folds-for-cross-validation" class="level1" data-number="s">
<h1 data-number="s"><span class="header-section-number">s</span> k_crossval: The number of folds for cross-validation</h1>
</section>
<section id="cluster-the-number-of-clusters-to-be-used-for-parallel-processing" class="level1" data-number="t">
<h1 data-number="t"><span class="header-section-number">t</span> cluster: The number of clusters to be used for parallel processing</h1>
</section>
<section id="alpha-the-significance-level-for-conditional-independence-tests-1" class="level1" data-number="u">
<h1 data-number="u"><span class="header-section-number">u</span> alpha: The significance level for conditional independence tests</h1>
</section>
<section id="ridge-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" class="level1" data-number="v">
<h1 data-number="v"><span class="header-section-number">v</span> ridge: A logical value indicating whether to use a custom threshold for network visualization</h1>
</section>
<section id="create-a-function-to-train-the-model-with-n-fold-cross-validation" class="level1" data-number="w">
<h1 data-number="w"><span class="header-section-number">w</span> Create a function to train the model with n fold cross validation,</h1>
</section>
<section id="n-train-and-100n-test" class="level1" data-number="x">
<h1 data-number="x"><span class="header-section-number">x</span> 100-100/n% train and 100/n% test</h1>
</section>
<section id="hold-out-each-partition-iteratively-so-that-no-data-are-totally-held-out" class="level1" data-number="y">
<h1 data-number="y"><span class="header-section-number">y</span> Hold out each partition iteratively so that no data are totally held out</h1>
<p>xval.the.model = function(data, k_crossval = 10, cluster, alpha, ridge) { # Start xval.the.model</p>
<p># Note: In cases of small sample sizes, k may need to be lowered</p>
<p># Store number of observations n &lt;- nrow(data)</p>
<p># Create a numeric vector named predcor initialized with a length # equal to the number of elements in the traits vector # This vector will be used to store correlation values between # predicted and observed values for each trait predcor &lt;- numeric(length(traits)) names(predcor) &lt;- traits</p>
<p># Repeat, creating a postcor vector that will be used to store # correlation values for posterior estimates postcor &lt;- numeric(length(traits)) names(postcor) &lt;- traits</p>
<p># Shuffle the data to get unbiased splits kcv &lt;- split(sample(n), seq_len(k_crossval)) # Store the length of each test set kcv.length &lt;- sapply(kcv, length)</p>
<p># Train network models and support the optional use of ridge regression for model re-fitting under certain conditions predicted &lt;- parLapply(kcv, cl = cluster, function(test) { # Start predicted</p>
<pre><code># Create a matrix to store the predicted values
pred &lt;- matrix(0, nrow = length(test), ncol = length(traits))
colnames(pred) &lt;- traits
# Create a matrix to store posterior estimates
post = matrix(0, nrow = length(test), ncol = length(traits))
colnames(post) = traits

# Print a message for each cross-validation fold
cat("* beginning cross-validation fold.\n")

# Split training and test data
dtraining &lt;- data[-test, ]
dtest &lt;- data[test, ]

# Fit the model on the training data (using function created in chunk above)
model &lt;- fit.the.model(dtraining, alpha = alpha)
fitted &lt;- bn.fit(model, dtraining[, nodes(model)])
# Optional re-fit with ridge regression
if (ridge) { # Start if
  
  # Load penalized library
  library(penalized)
  
  # Run ridge 
  for (no in nodes(fitted)) { # Start for
    
    node.parents &lt;- parents(fitted, no)
    
    if (length(node.parents) &lt; 3)
      next
    
    opt.lambda &lt;- optL2(response = dtraining[, no],
                        penalized = dtraining[, node.parents],
                        model = "linear", trace = FALSE,
                        minlambda2 = 10e-5, maxlambda = 500)$lambda
    fitted[[no]] &lt;- penalized(response = dtraining[, no],
                              penalized = dtraining[, node.parents],
                              model = "linear", trace = FALSE,
                              lambda1 = 0, lambda2 = opt.lambda)
    
  } # End for
  
} # End if -&gt; then 

# Subset the test data
dtest &lt;- dtest[, nodes(model)]

# Print message reporting the number of nodes in the Bayesian network model
cat("  &gt; model has", length(nodes(model)), "nodes.\n")

# Predict each trait in turn, given all the parents
# For each trait, use the fitted Bayesian network model (fitted) to make 
# predictions (predict) based on the test data (dtest) for the nodes
# in the model
# The predictions are stored in the pred matrix, where each column corresponds to a trait
for (t in traits)
  pred[, t] &lt;- predict(fitted, node = t, data = dtest[, nodes(model)])

# Iterate over the rows of the test data (dtest)
# For each row, calculate the posterior estimates for all traits (traits) 
# in the Bayesian network
# Note that this uses the cpdist function to estimate the conditional
# probability distribution of each trait given the evidence (values 
# of other variables) from the test data
# The resulting posterior estimates are stored in the post matrix, 
# with each column representing a trait
for (i in seq(nrow(dtest)))
  post[i, traits] = colMeans(cpdist(fitted, nodes = traits,
                                    evidence = as.list(dtest[i, names(dtest) %in% genes]),
                                    method = "lw", n = 1000))

# Return a list with the pred and post results
return(list(model = fitted, pred = pred, post = post))</code></pre>
<p>}) # End predicted</p>
<p># Summarize results from cross-validation and calculate and print statistics # on the model’s performance posterior &lt;- do.call(rbind, lapply(predicted, <code>[[</code>, “post”)) causal &lt;- do.call(rbind, lapply(predicted, <code>[[</code>, “pred”)) cat(“* overall cross-validated correlations:”)</p>
<p># Iterate over each trait in the traits vector to calculate # and print performance statistics for each trait separately for (t in traits) { predcor[t] = cor(causal[, t], data[unlist(kcv), t]) cat(” &gt; PREDCOR(“, t,”):“, predcor[t],”“) postcor[t] = cor(posterior[, t], data[unlist(kcv), t]) cat(” &gt; POSTCOR(“, t,”):“, postcor[t],”“) } # End for</p>
<p># Construct and return a list that summarizes various aspects of the # cross-validation results and model performance return(list(predicted = causal, posterior = posterior, observed = data[unlist(kcv), t], predcor = predcor, postcor = postcor, models = lapply(predicted, <code>[[</code>, “model”)))</p>
<p>} # End xval.the.model</p>
<section id="section-3" class="level51">
<p class="heading"></p>
</section>
<section id="section-4" class="level51">
<p class="heading"></p>
</section>
<section id="function-3-run_plot_graph" class="level2" data-number="y.1">
<h2 data-number="y.1" class="anchored" data-anchor-id="function-3-run_plot_graph"><span class="header-section-number">y.1</span> FUNCTION 3: run_plot_graph</h2>
</section>
</section>
<section id="this-code-defines-an-r-function-called-run_plot_graph-that-calling-the-functions" class="level1" data-number="z">
<h1 data-number="z"><span class="header-section-number">z</span> This code defines an R function called <code>run_plot_graph</code> that, calling the functions</h1>
</section>
<section id="above-learns-and-visualizes-the-bayesian-networks-in-the-data-including-customizing" class="level1" data-number="{">
<h1 data-number="{"><span class="header-section-number">{</span> above, learns and visualizes the Bayesian networks in the data, including customizing</h1>
</section>
<section id="node-labels-and-appearance-and-then-returns-results-related-to-the-network-analysis." class="level1" data-number="|">
<h1 data-number="|"><span class="header-section-number">|</span> node labels and appearance, and then returns results related to the network analysis.</h1>
</section>
<section id="note-that-the-following-function-uses-parallel-computing-which-involves-breaking-down" class="level1" data-number="}">
<h1 data-number="}"><span class="header-section-number">}</span> Note that the following function uses parallel computing which involves breaking down</h1>
</section>
<section id="a-complex-task-into-smaller-subtasks-that-can-be-executed-simultaneously" class="level1" data-number="~">
<h1 data-number="~"><span class="header-section-number">~</span> a complex task into smaller subtasks that can be executed simultaneously,</h1>
</section>
<section id="or-in-parallel-by-multiple-processors-or-computers-to-speed-up-computationally" class="level1" data-number="">
<h1 data-number=""><span class="header-section-number"></span> or in parallel, by multiple processors or computers to speed up computationally</h1>
</section>
<section id="intense-calculations." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> intense calculations.</h1>
</section>
<section id="data-the-dataset-to-be-used-for-plotting-the-networks" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> data: The dataset to be used for plotting the networks</h1>
</section>
<section id="k_crossval-the-number-of-folds-for-cross-validation-1" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> k_crossval: The number of folds for cross-validation</h1>
</section>
<section id="k_iterations-the-number-of-times-the-entire-process-which-includes-k-fold-cross-validation-is-repeated." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> k_iterations: The number of times the entire process (which includes k-fold cross-validation) is repeated.</h1>
</section>
<section id="alpha-the-significance-level-for-conditional-independence-tests-2" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> alpha: The significance level for conditional independence tests</h1>
</section>
<section id="use.custom.threshold-a-logical-value-indicating-whether-to-use-a-custom-threshold-for-network-visualization" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> use.custom.threshold: A logical value indicating whether to use a custom threshold for network visualization</h1>
</section>
<section id="custom.threshold-the-custom-threshold-value-to-be-used-if-use.custom.threshold-is-set-to-true" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> custom.threshold: The custom threshold value to be used if use.custom.threshold is set to TRUE</h1>
</section>
<section id="ncluster-the-number-of-clusters-for-parallel-computation" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> ncluster: The number of clusters for parallel computation</h1>
</section>
<section id="additional-arguments-to-be-passed-to-other-functions" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> …: Additional arguments to be passed to other functions</h1>
</section>
<section id="define-function-to-plot-the-networks" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Define function to plot the networks</h1>
<p>run_plot_graph &lt;- function(data = df, k_crossval = 10, k_iterations = 5, alpha = 0.01, use.custom.threshold = FALSE, custom.threshold = 0.90, ncluster = 8, custom_labels = custom_labels, …) { # Start run_plot_graph</p>
<p>#### CLUSTER # Create a cluster (cl) for parallel computation cl &lt;- makeCluster(ncluster)</p>
<p># Load library(bnlearn) and library(lme4) in the cluster invisible(clusterEvalQ(cl, library(bnlearn))) invisible(clusterEvalQ(cl, library(lme4)))</p>
<p># Export the traits, genes, and fit.the.model objects to the cluster # to make them available in the parallel environment clusterExport(cl = cl, c(“traits”, “genes”, “fit.the.model”))</p>
<p># Initialize the pr001 variable as a vector of length k with the mode set to “list” pr001 &lt;- vector(k_iterations, mode = “list”)</p>
<p># Perform cross-validation (xval.the.model) k times and store the results in the pr001 list for (i in seq_along(pr001)) { # Start for pr001[[i]] &lt;- xval.the.model(data, cluster = cl, alpha = alpha, ridge = FALSE, k_crossval = k_crossval) } # End for</p>
<p># Stop the cluster stopCluster(cl) #### CLUSTER</p>
<p># Calculate the average prediction correlation (pred.summary) and average # posterior correlation (post.summary) based on cross-validation results pred.summary &lt;- sapply(pr001, <code>[[</code>, “predcor”) print(rowMeans(pred.summary)) post.summary &lt;- sapply(pr001, <code>[[</code>, “postcor”) print(rowMeans(post.summary))</p>
<p># Average the network structures # Initialize a list (arclist) to store the network structures arclist &lt;- list() # Extract the models from pr001 and append their arcs to the arclist for (i in seq_along(pr001)) { # Start for # Extract the models run &lt;- pr001[[i]]$models for (j in seq_along(run)) arclist[[length(arclist) + 1]] = arcs(run[[j]]) } # End for # Compute the arc strengths # The nodes (genes and traits) present in the network structures are # identified and stored in the nodes variable nodes &lt;- unique(unlist(arclist))</p>
<p># Blacklist # Note: default boot.strength() or custom.strength() runs with # “cpdag = TRUE” which means reversible arcs can have positive # strength in both directions (so blacklist won’t work as we expect) # Use “cpdag = FALSE” to truly make the blacklist work as expected strength &lt;- custom.strength(arclist, nodes = nodes, cpdag = FALSE) # Estimate the threshold and average the networks averaged &lt;- averaged.network(strength)</p>
<p># Select relevant nodes (those with a degree greater than 0) # and store in relevant.nodes relevant.nodes &lt;- nodes(averaged)[sapply(nodes, degree, object = averaged) &gt; 0] # Create a subgraph (averaged2) based on the relevant nodes averaged2 &lt;- subgraph(averaged, relevant.nodes) # Extract the arc strengths (strength2) between relevant nodes strength2 &lt;- strength[(strength<span class="math inline">\(from %in% relevant.nodes) &amp; (strength\)</span>to %in% relevant.nodes), ]</p>
<p># Print information about the threshold value and minimum strength greater than the threshold cat(“threshold:”, attr(strength2, “threshold”), “”)</p>
<p># Customize node labels # Store the current node names from averaged current_nodes &lt;- nodes(averaged) # Create a custom labels dictionary with all nodes from averaged2 c.labels &lt;- custom_labels2[current_nodes] # Relabel nodes in averaged2 using the custom labels nodes(averaged2) &lt;- c.labels[nodes(averaged2)]</p>
<p># Update the ‘from’ and ‘to’ columns in the strength object with the custom labels strength2<span class="math inline">\(from &lt;- c.labels[strength2\)</span>from] strength2<span class="math inline">\(to &lt;- c.labels[strength2\)</span>to]</p>
<p># Update the attribute names in the strength and strength2 objects attr(strength2, “nodes”) &lt;- c.labels[attr(strength2, “nodes”)] attr(strength, “nodes”) &lt;- c.labels[attr(strength, “nodes”)] # Newly added to correct bug, 12-13-2023</p>
<p># Need to also updated results<span class="math inline">\(averaged and results\)</span>traits nodes(averaged) &lt;- c.labels[nodes(averaged)] traits &lt;- c.labels[traits]</p>
<p># strength2 retains nodes that are not in relevant.nodes as attributes # Since the plotting code below uses the attributes of strength2 and averaged2, # we need to update the attributes to reflect only relevant nodes # Extract unique nodes from both “from” and “to” columns nodes_from_strength &lt;- unique(c(strength<span class="math inline">\(from, strength\)</span>to)) nodes_from_strength2 &lt;- unique(c(strength2<span class="math inline">\(from, strength2\)</span>to)) # Create a list of nodes to keep (nodes that appear in either data frame) nodes_to_keep &lt;- intersect(nodes_from_strength, nodes_from_strength2) # Update the attributes of both data frames to only include nodes_to_keep attr(strength2, “nodes”) &lt;- attr(strength2, “nodes”)[nodes_to_keep]</p>
<p># Print the minimum strength observed that is &gt; threshold t &lt;- attr(strength2, “threshold”) v &lt;- strength2<span class="math inline">\(strength
  cat("min strength &gt; threshold: ", min(v[v &gt; t]), "\n")
  cat("strength: ", sort(strength2\)</span>strength))</p>
<p># Plot the network # The appearance of nodes and edges is customized based on the data if (use.custom.threshold) { cat(“Using custom threshold of”, custom.threshold, “”) gR = strength.plot( averaged2, strength2, shape = “rectangle”, layout = “fdp”, render = F, threshold = custom.threshold ) } else { gR = strength.plot( averaged2, strength2, shape = “rectangle”, layout = “fdp”, render = F ) }</p>
<p># Customize plot nodeRenderInfo(gR)<span class="math inline">\(fill = "lightblue"
  nodeRenderInfo(gR)\)</span>col = “darkblue” nodeRenderInfo(gR)<span class="math inline">\(fill[traits] = "limegreen"
  nodeRenderInfo(gR)\)</span>col[traits] = “darkgreen” a &lt;- arcs(bnlearn::subgraph(averaged, traits)) a &lt;- as.character(interaction(a[, “from”], a[, “to”], sep = “~”)) edgeRenderInfo(gR)<span class="math inline">\(col = "grey"
  edgeRenderInfo(gR)\)</span>col[a] = “darkgreen”</p>
<p># Render plot (gR) using Rgraphviz::renderGraph renderGraph(gR)</p>
<p># Create a list named results that contains the averaged2 network, # strength2, averaged network, traits variables, and threshold value results &lt;- list( averaged2 = averaged2, strength2 = strength2, averaged = averaged, traits = traits, threshold = t )</p>
<p># Return results return(results)</p>
<p>} # End run_plot_graph</p>
<section id="section-5" class="level51">
<p class="heading"></p>
</section>
<section id="section-6" class="level51">
<p class="heading"></p>
</section>
<section id="function-4-match.arcs.and.directions" class="level2" data-number="�.1">
<h2 data-number="�.1" class="anchored" data-anchor-id="function-4-match.arcs.and.directions"><span class="header-section-number">�.1</span> FUNCTION 4: match.arcs.and.directions</h2>
</section>
</section>
<section id="this-code-defines-an-r-function-called-match.arcs.and.directions-which-is" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> This code defines an R function called <code>match.arcs.and.directions</code> which is</h1>
</section>
<section id="designed-to-match-arcs-in-a-network-with-their-corresponding-direction" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> designed to match arcs in a network with their corresponding direction</h1>
</section>
<section id="coefficients-in-a-strengths-data-frame-providing-flexibility-to-either" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> coefficients in a strengths data frame, providing flexibility to either</h1>
</section>
<section id="keep-or-discard-additional-information-based-on-the-keep-argument." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> keep or discard additional information based on the keep argument.</h1>
</section>
<section id="arcs-arcs-within-a-bayesian-network-representing-relationships" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> arcs: Arcs within a Bayesian network representing relationships</h1>
</section>
<section id="between-nodes-genes-or-traits" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> between nodes (genes or traits)</h1>
</section>
<section id="nodes-nodes-within-a-bayesian-network-which-can-represent" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> nodes: Nodes within a Bayesian network, which can represent</h1>
</section>
<section id="either-genes-or-traits" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> either genes or traits</h1>
</section>
<section id="strengths-strength-coefficients-associated-with-the-arcs-in" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> strengths: Strength coefficients associated with the arcs in</h1>
</section>
<section id="the-bayesian-network-indicating-the-strength-and-direction" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> the Bayesian network, indicating the strength and direction</h1>
</section>
<section id="of-relationships" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> of relationships</h1>
</section>
<section id="keep-a-logical-value-indicating-whether-to-keep-additional" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> keep: A logical value indicating whether to keep additional</h1>
</section>
<section id="information-when-matching-arcs-and-directions-in-the" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> information when matching arcs and directions in the</h1>
</section>
<section id="bayesian-network" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Bayesian network</h1>
</section>
<section id="define-function-to-match-arcs-and-directions" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Define function to match arcs and directions</h1>
<p>match.arcs.and.directions &lt;- function (arcs, nodes, strengths, keep = FALSE) { # Start match.arcs.and.directions if (nrow(strengths) &lt; nrow(arcs)) stop(“insufficient number of strength coefficients.”) a_hash &lt;- interaction(arcs[, “from”], arcs[, “to”]) s_hash &lt;- interaction(strengths[, “from”], strengths[, “to”]) if (keep) { s &lt;- strengths[match(a_hash, s_hash), , drop = FALSE] coef = s$direction } else { s &lt;- strengths[match(a_hash, s_hash), “direction”] from &lt;- strengths[match(a_hash, s_hash), “from”] to &lt;- strengths[match(a_hash, s_hash), “to”] names(s) &lt;- paste0(from, “~”, to) coef &lt;- s } if (any(is.na(coef))) { missing = apply(arcs[is.na(coef), , drop = FALSE], 1, function(x) { paste(” (“, x[1],”, “, x[2],”)“, sep =”“) }) stop(”the following arcs do not have a corresponding direction coefficients:“, missing,”.” ) }</p>
<p># Return return(s) } # End match.arcs.and.directions</p>
<section id="section-7" class="level51">
<p class="heading"></p>
</section>
<section id="section-8" class="level51">
<p class="heading"></p>
</section>
<section id="function-5-redraw.graph.labels" class="level2" data-number="�.1">
<h2 data-number="�.1" class="anchored" data-anchor-id="function-5-redraw.graph.labels"><span class="header-section-number">�.1</span> FUNCTION 5: redraw.graph.labels</h2>
</section>
</section>
<section id="this-code-defines-the-redraw.graph.labels-function-which-enhances-the-graph" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> This code defines the <code>redraw.graph.labels</code> function, which enhances the graph</h1>
</section>
<section id="created-above-by-adding-labels-to-its-edges.-these-labels-make-the-graph-more" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> created above by adding labels to its edges. These labels make the graph more</h1>
</section>
<section id="informative-by-conveying-information-about-the-strength-and-direction-of" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> informative by conveying information about the strength and direction of</h1>
</section>
<section id="relationships-represented-by-each-edge-in-the-graph." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> relationships represented by each edge in the graph.</h1>
</section>
<section id="strength-a-measure-of-confidence-of-that-edge-while-fixing-the-rest-of" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Strength: a measure of confidence of that edge while fixing the rest of</h1>
</section>
<section id="the-network-structure-and-is-defined-as-the-empirical-frequency-a-specific" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> the network structure and is defined as the empirical frequency a specific</h1>
</section>
<section id="edge-is-observed-over-a-set-of-networks-learned-from-iterations" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> edge is observed over a set of networks learned from iterations</h1>
</section>
<section id="i.e.-the-number-of-times-the-edge-was-present-out-of-the-total-number-of" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> (i.e., the number of times the edge was present out of the total number of</h1>
</section>
<section id="iterations." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> iterations).</h1>
</section>
<section id="direction-the-probability-of-the-edges-direction-conditional-on-the-edges" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Direction: the probability of the edge’s direction conditional on the edge’s</h1>
</section>
<section id="presence-within-the-network-i.e.-the-number-of-times-the-edge-traveled-in" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> presence within the network (i.e., the number of times the edge traveled in</h1>
</section>
<section id="a-specific-direction-out-of-the-total-number-of-iterations" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> a specific direction out of the total number of iterations</h1>
</section>
<section id="in-which-it-was-present." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> in which it was present).</h1>
</section>
<section id="averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> averaged2: Bayesian network learned through functions above (relevant nodes only)</h1>
</section>
<section id="strength2-bayesian-network-strengths-learned-through-the-functions-above" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> strength2: Bayesian network strengths learned through the functions above</h1>
</section>
<section id="averaged-bayesian-network-learned-through-functions-above-all-nodes" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> averaged: Bayesian network learned through functions above (all nodes)</h1>
</section>
<section id="traits-phenotypes-of-interest" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> traits: Phenotypes of interest</h1>
</section>
<section id="custom.threshold-custom-threshold-to-customize-line-thicknesstype" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> custom.threshold: Custom threshold to customize line thickness/type</h1>
</section>
<section id="define-function-to-redraw-graph-labels" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Define function to redraw graph labels</h1>
<p>redraw.graph.labels &lt;- function(averaged2, strength2, averaged, traits, custom.threshold) { # Start redraw.graph.labels gR &lt;- strength.plot( averaged2, strength2, shape = “rectangle”, layout = “fdp”, render = F, threshold = custom.threshold ) x &lt;- averaged2 str &lt;- match.arcs.and.directions( arcs = x<span class="math inline">\(arcs,
        nodes = names(x\)</span>nodes), strengths = strength2 ) str2 &lt;- bnlearn:::match.arcs.and.strengths( arcs = x<span class="math inline">\(arcs,
      nodes = names(x\)</span>nodes), strengths = strength2 )</p>
<pre><code># Customize edge labels 
labels2 &lt;- paste0(round(str2, 2), ":", round(str, 2))
names(labels2) &lt;- names(str)

# Begin plot
gR &lt;- Rgraphviz::layoutGraph(gR,
                             edgeAttrs = list(label = labels2),
                             layoutType = "fdp")

# Customize plot
nodeRenderInfo(gR)$fill[genes] = "grey"
nodeRenderInfo(gR)$fill = "lightblue"
nodeRenderInfo(gR)$col = "darkblue"
nodeRenderInfo(gR)$fill[traits] = "limegreen"
nodeRenderInfo(gR)$col[traits] = "darkgreen"
a = arcs(bnlearn::subgraph(averaged2, traits))
a = as.character(interaction(a[, "from"], a[, "to"], sep = "~"))
edgeRenderInfo(gR)$col = "grey"
edgeRenderInfo(gR)$col[a] = "darkgreen"
Rgraphviz::renderGraph(gR)</code></pre>
<p>} # End redraw.graph.labels</p>
<section id="section-9" class="level51">
<p class="heading"></p>
</section>
<section id="section-10" class="level51">
<p class="heading"></p>
</section>
<section id="function-6-redraw.label.ggnet" class="level2" data-number="�.1">
<h2 data-number="�.1" class="anchored" data-anchor-id="function-6-redraw.label.ggnet"><span class="header-section-number">�.1</span> FUNCTION 6: redraw.label.ggnet</h2>
</section>
</section>
<section id="the-following-function-redraw.label.ggnet-is-used-for-an-additional" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> The following function, <code>redraw.label.ggnet</code>, is used for an additional</h1>
</section>
<section id="redrawing-of-the-bayesian-network-but-in-a-ggnetworkggplot2-framework" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> redrawing of the Bayesian network but in a ggnetwork/ggplot2 framework,</h1>
</section>
<section id="enhancing-it-with-labels-and-color-coding-for-interpretation-in-the-context" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> enhancing it with labels and color-coding for interpretation in the context</h1>
</section>
<section id="of-bayesian-analysis.-of-note-this-function-calls-in-ggrepel-so" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> of Bayesian analysis. Of note, this function calls in <code>ggrepel()</code> so</h1>
</section>
<section id="that-node-and-edge-labels-do-not-overlap-further-supporting-the-user" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> that node and edge labels do not overlap, further supporting the user</h1>
</section>
<section id="in-creating-publication-quality-figures." class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> in creating publication quality figures.</h1>
</section>
<section id="averaged2-bayesian-network-learned-through-functions-above-relevant-nodes-only-1" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> averaged2: Bayesian network learned through functions above (relevant nodes only)</h1>
</section>
<section id="strength2-bayesian-network-strengths-learned-through-the-functions-above-1" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> strength2: Bayesian network strengths learned through the functions above</h1>
</section>
<section id="averaged-bayesian-network-learned-through-functions-above-all-nodes-1" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> averaged: Bayesian network learned through functions above (all nodes)</h1>
</section>
<section id="traits-phenotype-of-interest" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> traits: Phenotype of interest</h1>
</section>
<section id="additional-arguments-to-be-passed-to-other-functions-1" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> …: Additional arguments to be passed to other functions</h1>
</section>
<section id="define-function-to-redraw-the-graph-in-ggnetworkggplot2-framework" class="level1" data-number="�">
<h1 data-number="�"><span class="header-section-number">�</span> Define function to redraw the graph in ggnetwork/ggplot2 framework</h1>
<p>redraw.label.ggnet &lt;- function(averaged2, strength2, averaged, traits, df_vertex_table=df_vertex_table, …) { # Start redraw.label.ggnet</p>
<p>x &lt;- averaged2</p>
<p># Relies on bnlearn code to return a directed graph # Get the directions and strengths vec_dir &lt;- match.arcs.and.directions( arcs = x<span class="math inline">\(arcs,
    nodes = names(x\)</span>nodes), strengths = strength2 ) vec_str &lt;- bnlearn:::match.arcs.and.strengths( arcs = x<span class="math inline">\(arcs,
    nodes = names(x\)</span>nodes), strengths = strength2 )</p>
<p># Create data.frame with node/directions/strengths # Join this to the data.frame created by ggnetwork df_dir_str &lt;- data.frame(do.call(rbind, str_split(names(vec_dir), “~”)), vec_dir, vec_str, stringsAsFactors = F) names(df_dir_str)</p>
<p># Convert NEL graph to iGraph to ggnetwork data.frame # This step ‘shuffles’ the edge orders so vec_dir and vec_str are not in same order # as df_ggnetwork graph_igraph &lt;- igraph::igraph.from.graphNEL(as.graphNEL(averaged2)) df_ggnetwork &lt;- ggnetwork(graph_igraph, …) df_ggnetwork<span class="math inline">\(vertex.names &lt;- as.character(df_ggnetwork\)</span>name) df_ggnetwork<span class="math inline">\(xend &lt;- as.numeric(df_ggnetwork\)</span>xend) df_ggnetwork<span class="math inline">\(yend &lt;- as.numeric(df_ggnetwork\)</span>yend) df_ggnetwork<span class="math inline">\(x &lt;- as.numeric(df_ggnetwork\)</span>x) df_ggnetwork<span class="math inline">\(y &lt;- as.numeric(df_ggnetwork\)</span>y)</p>
<p># Create a data.frame of node coordinates as this is needed to determine the # destination node/end vertex df_node_coord &lt;- df_ggnetwork %&gt;% filter(is.na(weight)) %&gt;% select(vertex.names, x, y) %&gt;% transmute(vertex.names.end = vertex.names, xend = x, yend = y)</p>
<p># Join back to form end vertex # This step relies on fuzzyjoin with L2 norm &lt; 0.05 since ggnetwork shifts the node # positions around randomly df_ggnetwork &lt;- df_ggnetwork %&gt;% fuzzyjoin::distance_left_join(df_node_coord, by = c(“xend”, “yend”), max_dist = 0.05)</p>
<p># Join the direction/probability table based on source node and end node df_ggnetwork &lt;- df_ggnetwork %&gt;% left_join(df_dir_str, by = c(“vertex.names” = “X1”, “vertex.names.end” = “X2”))</p>
<p># Join node type table df_ggnetwork &lt;- df_ggnetwork %&gt;% left_join(df_vertex_table, by = “vertex.names”)</p>
<p># Create a column for node colors based on the node type # Note that in this example, we only have two node types (symptom/SNP) df_ggnetwork &lt;- df_ggnetwork %&gt;% mutate(node_color = ifelse(grepl(“^rs”, vertex.names), “tan3”, “steelblue”))</p>
<p># Fix the data.frame column names, fuzzyjoin appended .x and .y to column names names(df_ggnetwork)[names(df_ggnetwork) == “xend.x”] &lt;- “xend” names(df_ggnetwork)[names(df_ggnetwork) == “yend.x”] &lt;- “yend”</p>
<p># Plot with ggplot and ggnetwork geoms ggplot(df_ggnetwork, aes( x = x, y = y, xend = xend, yend = yend)) + geom_edges(aes(linetype = (vec_str &gt; 0.9)), color = “grey50”, arrow = arrow(length = unit(8, “pt”), type = “open”)) + geom_nodes(size = 10, aes(shape = type, color = node_color, fill = node_color), stroke = 0.5, # Set the size of the border color = “black”) + # Set the color of the border geom_edgelabel_repel(aes(label = paste0( round(vec_str, 2), “:”, round(vec_dir, 2)))) + geom_nodelabel_repel( aes(label = vertex.names), force = 1, fontface = “bold”, box.padding = unit(1.2, “lines”), color = “black”) + scale_color_manual(values = c(“tan3” = “tan3”, “steelblue” = “steelblue”)) + scale_fill_manual(values = c(“tan3” = “tan3”, “steelblue” = “steelblue”)) + scale_linetype_manual( “Node Signif.”, values = c(“FALSE” = “dotted”, “TRUE” = “solid”), label = c(“&lt;= 90 Strength”, “&gt; 0.90 Strength”)) + scale_shape_manual(values = c(21:23)) + theme_blank(legend.position = “none”)</p>
<p>} # End redraw.label.ggnet</p>
<section id="section-11" class="level51">
<p class="heading"></p>


</section>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/lwheinsberg\.github\.io\/mvNUR");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./mvbimbam.html" class="pagination-link  aria-label=" &lt;span="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">mvBIMBAM</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>